<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create comparisons between DNA segments — comparisons_from_dna_segs • genoPlotR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create comparisons between DNA segments — comparisons_from_dna_segs"><meta name="description" content="Create a list of comparison objects from a list of dna_seg objects or
files by parsing (and executing) sequence alignments. If these files already
exist, then those will be parsed. If not, DIAMOND or a BLAST program can be
executed to generate the sequence alignment results between the dna_seg,
with respect to the order of dna_segs. Executing DIAMOND or BLAST requires
that the command-line implementations of these tools are installed."><meta property="og:description" content="Create a list of comparison objects from a list of dna_seg objects or
files by parsing (and executing) sequence alignments. If these files already
exist, then those will be parsed. If not, DIAMOND or a BLAST program can be
executed to generate the sequence alignment results between the dna_seg,
with respect to the order of dna_segs. Executing DIAMOND or BLAST requires
that the command-line implementations of these tools are installed."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genoPlotR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/genoPlotR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/wrapper_script.html">Running genoPlotR from a command-line interface</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create comparisons between DNA segments</h1>

      <div class="d-none name"><code>comparisons_from_dna_segs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Create a list of <code>comparison</code> objects from a list of <code>dna_seg</code> objects or
files by parsing (and executing) sequence alignments. If these files already
exist, then those will be parsed. If not, DIAMOND or a BLAST program can be
executed to generate the sequence alignment results between the <code>dna_seg</code>,
with respect to the order of <code>dna_segs</code>. Executing DIAMOND or BLAST requires
that the command-line implementations of these tools are installed.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">comparisons_from_dna_segs</span><span class="op">(</span></span>
<span>  dna_segs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seg_labels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  files <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>  tool <span class="op">=</span> <span class="st">"blast"</span>,</span>
<span>  algorithm <span class="op">=</span> <span class="st">"blastp"</span>,</span>
<span>  sensitivity <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>  output_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  all_vs_all <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  filt_high_evalue <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filt_low_per_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filt_length <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  use_cache <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dna-segs">dna_segs<a class="anchor" aria-label="anchor" href="#arg-dna-segs"></a></dt>
<dd><p>A list of <code>dna_seg</code> objects to create comparisons between.
Either <code>dna_segs</code> or <code>files</code> must be provided.</p></dd>


<dt id="arg-seg-labels">seg_labels<a class="anchor" aria-label="anchor" href="#arg-seg-labels"></a></dt>
<dd><p>A character vector containing DNA segment labels.</p></dd>


<dt id="arg-files">files<a class="anchor" aria-label="anchor" href="#arg-files"></a></dt>
<dd><p>A character vector, containing file paths to the FASTA or
GenBank files. The comparisons will be made between these files. Either
<code>dna_segs</code> or <code>files</code> must be provided.</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>Determines how the comparisons will be filtered.
<code>"besthit"</code>, <code>"bidirectional"</code>, or <code>"full"</code>. If mode is <code>"besthit"</code>, only the
best hit will be taken from each input query (see <a href="best_hit.html">best_hit</a>). If mode is
<code>"bidirectional"</code>, then hits are only kept if they are the best hits for
their query in both directions (see <a href="bidirectional_best_hit.html">bidirectional_best_hit</a>). <code>"full"</code> means
that all sequence alignment results are considered.</p></dd>


<dt id="arg-tool">tool<a class="anchor" aria-label="anchor" href="#arg-tool"></a></dt>
<dd><p>Choice of sequence alignment tool. Either <code>"blast"</code> or
<code>"diamond"</code>.</p></dd>


<dt id="arg-algorithm">algorithm<a class="anchor" aria-label="anchor" href="#arg-algorithm"></a></dt>
<dd><p>Choice of BLAST algorithm to run. One of: <code>"blastp"</code>,
<code>"blastp-fast"</code>, <code>"blastp-short"</code>, <code>"tblastx"</code>, <code>"blastn"</code>, <code>"blastn-short"</code>,
<code>"megablast"</code>, or <code>"dc-megablast"</code>.</p></dd>


<dt id="arg-sensitivity">sensitivity<a class="anchor" aria-label="anchor" href="#arg-sensitivity"></a></dt>
<dd><p>Choice of sensitivity option when running DIAMOND. One of:
<code>"fast"</code>, <code>"default"</code>, <code>"mid-sensitive"</code>, <code>"sensitive"</code>, <code>"more-sensitive"</code>,
<code>"very-sensitive"</code>, or <code>"ultra-sensitive"</code>.</p></dd>


<dt id="arg-output-path">output_path<a class="anchor" aria-label="anchor" href="#arg-output-path"></a></dt>
<dd><p>Path to the folder that will contain the output files.
Both the sequence alignment result and the FASTA files used to make them will
be stored here.</p></dd>


<dt id="arg-all-vs-all">all_vs_all<a class="anchor" aria-label="anchor" href="#arg-all-vs-all"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, sequence alignments will be performed
for every combination of the inputs, instead of just the ones necessary for
plotting. Note that this can take a long time, so use with caution.</p></dd>


<dt id="arg-filt-high-evalue">filt_high_evalue<a class="anchor" aria-label="anchor" href="#arg-filt-high-evalue"></a></dt>
<dd><p>A numerical, filters out all comparisons with an
e-value higher than this value (unfiltered when left as <code>NULL</code>).</p></dd>


<dt id="arg-filt-low-per-id">filt_low_per_id<a class="anchor" aria-label="anchor" href="#arg-filt-low-per-id"></a></dt>
<dd><p>A numerical, filters out all comparisons with a
percentage identity lower than this value (unfiltered when left as <code>NULL</code>).</p></dd>


<dt id="arg-filt-length">filt_length<a class="anchor" aria-label="anchor" href="#arg-filt-length"></a></dt>
<dd><p>A number indicating the minimum length required for hits,
or <code>"auto"</code>. If <code>"auto"</code>, it will be determined based on the choice of <code>tool</code>
and <code>algorithm</code> (150 for DIAMOND or any blastp algorithm, 450 for tblastx,
900 for any blastn algorithm).</p></dd>


<dt id="arg-use-cache">use_cache<a class="anchor" aria-label="anchor" href="#arg-use-cache"></a></dt>
<dd><p>Logical. If <code>FALSE</code>, it will never check for existing files.
This includes the FASTA files used as input for sequcence alignment, the
database files used by DIAMOND and BLAST, and the sequence alignment results
themselves.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, reports timings when creating new files.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Arguments to pass to other functions (the functions executing
the sequence alignments tools, <a href="run_sequence_alignment.html">run_blast</a>, and <a href="run_sequence_alignment.html">run_diamond</a>).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of <code>comparison</code> objects.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Unless <code>use_cache</code> is set to <code>FALSE</code>, this function will look for the files
required using a combination of the <code>seg_labels</code> (if these are provided), and
the names of the <code>dna_segs</code> or <code>files</code> that were provided as input. If it
cannot find sequence alignment results in the form of <code>"query_subject"</code>
(or to put it differently, <code>"dna_seg1_dna_seg2"</code>), then it will run DIAMOND
or BLAST to generate these results. Using this system, it also looks for the
FASTA files required as input for the sequence alignment.</p>
<p>If <code>output_path</code> is left as <code>NULL</code>, the current working directory will be
used instead.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="run_sequence_alignment.html">run_blast</a>, <a href="run_sequence_alignment.html">run_diamond</a>, <a href="dna_seg_to_fasta.html">dna_seg_to_fasta</a>, <a href="genbank_to_fasta.html">genbank_to_fasta</a></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Mike Puijk</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co">## Comparisons from a vector of GenBank files using DIAMOND</span></span></span>
<span class="r-in"><span><span class="va">comparisons</span> <span class="op">&lt;-</span> <span class="fu">comparisons_from_dna_segs</span><span class="op">(</span></span></span>
<span class="r-in"><span>  files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genome1.gb"</span>, <span class="st">"genome2.gb"</span>, <span class="st">"genome2.gb"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  tool <span class="op">=</span> <span class="st">"diamond"</span>,</span></span>
<span class="r-in"><span>  output_path <span class="op">=</span> <span class="st">"output/diamond"</span>,</span></span>
<span class="r-in"><span>  sensitivity <span class="op">=</span> <span class="st">"very-sensitive"</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mike Puijk.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

