<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Lists gene_types for dna_seg objects — gene_types • genoPlotR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Lists gene_types for dna_seg objects — gene_types"><meta name="description" content="Returns a character vector containing the available gene types when plotting
dna_seg features."><meta property="og:description" content="Returns a character vector containing the available gene types when plotting
dna_seg features."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genoPlotR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/genoPlotR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/wrapper_script.html">Running genoPlotR from a command-line interface</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Lists gene_types for dna_seg objects</h1>

      <div class="d-none name"><code>gene_types.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Returns a character vector containing the available gene types when plotting
<code>dna_seg</code> features.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gene_types</span><span class="op">(</span>auto <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-auto">auto<a class="anchor" aria-label="anchor" href="#arg-auto"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, includes the <code>"auto"</code> gene type in the
output, which has <a href="plot_gene_map.html">plot_gene_map</a> determine what gene type to use
automatically.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A character vector.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>dna_segs</code> may contain the <code>gene_type</code> column, which determines the shape
of <code>dna_seg</code> features when they are plotted using <a href="plot_gene_map.html">plot_gene_map</a>.
Elements in this column should either be one of the predefined gene types
returned by this function, or they must refer to a graphical function with
exactly the same name that returns a <code>grob</code> or a <code>gList</code> object.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="plot_gene_map.html">plot_gene_map</a>, <a href="dna_seg.html">dna_seg</a></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Lionel Guy</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## To view pre-coded gene types:</span></span></span>
<span class="r-in"><span><span class="fu">gene_types</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "auto"            "arrows"          "headless_arrows" "blocks"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "bars"            "points"          "text"            "lines"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "side_blocks"     "side_bars"       "side_points"     "side_text"      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "side_lines"      "introns"         "exons"           "side_exons"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17] "boundaries"     </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Load data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">barto</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu">gene_types</span><span class="op">(</span>auto <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get a small subset from the barto dataset</span></span></span>
<span class="r-in"><span><span class="va">dna_seg</span> <span class="op">&lt;-</span> <span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">dna_seg</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="gene_types-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Change gene_types and plot again</span></span></span>
<span class="r-in"><span><span class="va">dna_seg</span><span class="op">$</span><span class="va">gene_type</span> <span class="op">&lt;-</span> <span class="fu">gene_types</span><span class="op">(</span>auto <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dna_seg</span><span class="op">$</span><span class="va">fill</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">rainbow</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dna_seg_r</span> <span class="op">&lt;-</span> <span class="va">dna_seg</span></span></span>
<span class="r-in"><span><span class="va">dna_seg_r</span><span class="op">$</span><span class="va">strand</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="va">dna_seg</span><span class="op">$</span><span class="va">strand</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Add an annotation</span></span></span>
<span class="r-in"><span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="annotation.html">annotation</a></span><span class="op">(</span><span class="fu"><a href="middle.html">middle</a></span><span class="op">(</span><span class="va">dna_seg</span><span class="op">)</span>, text <span class="op">=</span> <span class="va">dna_seg</span><span class="op">$</span><span class="va">gene_type</span>, rot <span class="op">=</span> <span class="fl">45</span>,</span></span>
<span class="r-in"><span>                    col <span class="op">=</span> <span class="va">dna_seg</span><span class="op">$</span><span class="va">col</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Plot</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">dna_seg</span>, <span class="va">dna_seg_r</span><span class="op">)</span>, annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">annot</span>, <span class="va">annot</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>              annotation_height <span class="op">=</span> <span class="fl">5</span>, dna_seg_line <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/gray.html" class="external-link">grey</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="gene_types-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mike Puijk.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

