<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ genoPlotR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genoPlotR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/genoPlotR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/wrapper_script.html">Running genoPlotR from a command-line interface</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="annotation.html">annotation()</a></code> <code><a href="annotation.html">as.annotation()</a></code> <code><a href="annotation.html">is.annotation()</a></code>

        </dt>
        <dd>annotation class and methods</dd>
      </dl><dl><dt>

          <code><a href="artemisColors.html">artemisColors()</a></code>

        </dt>
        <dd>Artemis Colors</dd>
      </dl><dl><dt>

          <code><a href="auto_annotate.html">auto_annotate()</a></code>

        </dt>
        <dd>Auto-annotate dna_segs</dd>
      </dl><dl><dt>

          <code><a href="barto.html">barto</a></code>

        </dt>
        <dd>Comparison of 4 Bartonella genomes</dd>
      </dl><dl><dt>

          <code><a href="best_hit.html">best_hit()</a></code>

        </dt>
        <dd>Filter a comparison to include only the best hits for each query</dd>
      </dl><dl><dt>

          <code><a href="bidirectional_best_hit.html">bidirectional_best_hit()</a></code>

        </dt>
        <dd>Filter a comparison to include only bidirectional best hits</dd>
      </dl><dl><dt>

          <code><a href="chrY_subseg.html">chrY_subseg</a></code>

        </dt>
        <dd>Comparisons of subsegments of the Y chromosome in human and chimp</dd>
      </dl><dl><dt>

          <code><a href="comparison.html">comparison()</a></code> <code><a href="comparison.html">as.comparison()</a></code> <code><a href="comparison.html">is.comparison()</a></code>

        </dt>
        <dd>comparison class and methods</dd>
      </dl><dl><dt>

          <code><a href="comparisons_from_dna_segs.html">comparisons_from_dna_segs()</a></code>

        </dt>
        <dd>Create comparisons between DNA segments</dd>
      </dl><dl><dt>

          <code><a href="dna_seg.html">dna_seg()</a></code> <code><a href="dna_seg.html">as.dna_seg()</a></code> <code><a href="dna_seg.html">is.dna_seg()</a></code>

        </dt>
        <dd>dna_seg class and methods</dd>
      </dl><dl><dt>

          <code><a href="dna_seg_to_fasta.html">dna_seg_to_fasta()</a></code>

        </dt>
        <dd>Convert dna_seg objects to FASTA files</dd>
      </dl><dl><dt>

          <code><a href="edit_dna_segs.html">edit_dna_segs()</a></code>

        </dt>
        <dd>Edit dna_seg features</dd>
      </dl><dl><dt>

          <code><a href="genbank_to_fasta.html">genbank_to_fasta()</a></code>

        </dt>
        <dd>Convert GenBank files to FASTA files</dd>
      </dl><dl><dt>

          <code><a href="gene_types.html">gene_types()</a></code>

        </dt>
        <dd>Lists gene_types for dna_seg objects</dd>
      </dl><dl><dt>

          <code><a href="gradient_color_scheme.html">gradient_color_scheme()</a></code> <code><a href="gradient_color_scheme.html">apply_color_scheme()</a></code>

        </dt>
        <dd>Generate a gradient color scheme</dd>
      </dl><dl><dt>

          <code><a href="human_nt.html">human_nt()</a></code>

        </dt>
        <dd>Human-readable nucleotide scale</dd>
      </dl><dl><dt>

          <code><a href="make_unique_ids.html">make_unique_ids()</a></code>

        </dt>
        <dd>Make unique IDs for dna_segs</dd>
      </dl><dl><dt>

          <code><a href="mauve_bbone.html">mauve_bbone</a></code>

        </dt>
        <dd>Mauve backbone of 4 Bartonella genomes</dd>
      </dl><dl><dt>

          <code><a href="max_by_group.html">max_by_group()</a></code>

        </dt>
        <dd>Filter dna_seg features by looking at a maximum within groups</dd>
      </dl><dl><dt>

          <code><a href="middle.html">middle()</a></code>

        </dt>
        <dd>Find the middle point of dna_seg features</dd>
      </dl><dl><dt>

          <code><a href="permute_dna_segs.html">permute_dna_segs()</a></code>

        </dt>
        <dd>Reorder dna_segs or labels to match a tree</dd>
      </dl><dl><dt>

          <code><a href="plot_gene_map.html">plot_gene_map()</a></code>

        </dt>
        <dd>Plot gene and genome maps</dd>
      </dl><dl><dt>

          <code><a href="print_comparison.html">print_comparison()</a></code>

        </dt>
        <dd>Print out a comparison</dd>
      </dl><dl><dt>

          <code><a href="range.html">range(<i>&lt;annotation&gt;</i>)</a></code> <code><a href="range.html">range(<i>&lt;comparison&gt;</i>)</a></code> <code><a href="range.html">range(<i>&lt;dna_seg&gt;</i>)</a></code>

        </dt>
        <dd>Range calculation</dd>
      </dl><dl><dt>

          <code><a href="read_comparison.html">read_comparison_from_file()</a></code> <code><a href="read_comparison.html">read_comparison_from_files()</a></code> <code><a href="read_comparison.html">read_comparison_from_blast()</a></code> <code><a href="read_comparison.html">read_comparison_from_tab()</a></code>

        </dt>
        <dd>Creating comparisons between genomic sequences from files</dd>
      </dl><dl><dt>

          <code><a href="read_dna_seg.html">read_dna_seg_from_file()</a></code> <code><a href="read_dna_seg.html">read_dna_seg_from_files()</a></code> <code><a href="read_dna_seg.html">read_dna_seg_from_embl()</a></code> <code><a href="read_dna_seg.html">read_dna_seg_from_genbank()</a></code> <code><a href="read_dna_seg.html">read_dna_seg_from_fasta()</a></code> <code><a href="read_dna_seg.html">read_dna_seg_from_ptt()</a></code> <code><a href="read_dna_seg.html">read_dna_seg_from_tab()</a></code>

        </dt>
        <dd>Creating dna_segs from files</dd>
      </dl><dl><dt>

          <code><a href="read_mauve_backbone.html">read_mauve_backbone()</a></code>

        </dt>
        <dd>Read mauve backbone files to create dna_segs and comparisons</dd>
      </dl><dl><dt>

          <code><a href="read_offsets.html">read_offsets()</a></code>

        </dt>
        <dd>Parse <code>dna_seg</code> offset values for plotting</dd>
      </dl><dl><dt>

          <code><a href="read_offsets_file.html">read_offsets_file()</a></code>

        </dt>
        <dd>Parse <code>dna_seg</code> offset values for plotting from a file</dd>
      </dl><dl><dt>

          <code><a href="read_orthogroup.html">read_orthogroup_from_file()</a></code> <code><a href="read_orthogroup.html">read_orthogroup_from_orthomcl()</a></code> <code><a href="read_orthogroup.html">read_orthogroup_from_orthofinder()</a></code> <code><a href="read_orthogroup.html">read_orthogroup_from_mmseqs2()</a></code> <code><a href="read_orthogroup.html">read_orthogroup_from_diamond()</a></code>

        </dt>
        <dd>Create comparisons between dna_segs by reading in a file of groupings</dd>
      </dl><dl><dt>

          <code><a href="read_xlims.html">read_xlims()</a></code>

        </dt>
        <dd>Parse <code>xlims</code>, a set of <code>dna_seg</code> positions for plotting</dd>
      </dl><dl><dt>

          <code><a href="read_xlims_file.html">read_xlims_file()</a></code>

        </dt>
        <dd>Parse <code>xlims</code> from a file, a set of <code>dna_seg</code> positions for plotting</dd>
      </dl><dl><dt>

          <code><a href="reverse.html">reverse()</a></code>

        </dt>
        <dd>Reverse objects</dd>
      </dl><dl><dt>

          <code><a href="run_sequence_alignment.html">run_blast()</a></code> <code><a href="run_sequence_alignment.html">run_diamond()</a></code>

        </dt>
        <dd>Perform sequence alignments using DIAMOND or BLAST</dd>
      </dl><dl><dt>

          <code><a href="seg_plot.html">seg_plot()</a></code> <code><a href="seg_plot.html">as.seg_plot()</a></code> <code><a href="seg_plot.html">is.seg_plot()</a></code>

        </dt>
        <dd>seg_plot class and methods</dd>
      </dl><dl><dt>

          <code><a href="sequential_color_scheme.html">sequential_color_scheme()</a></code>

        </dt>
        <dd>Apply a sequential color scheme to dna_segs and comparisons</dd>
      </dl><dl><dt>

          <code><a href="sequential_updates.html">sequential_updates()</a></code>

        </dt>
        <dd>Update dna_segs and comparisons sequentially in plotting order</dd>
      </dl><dl><dt>

          <code><a href="three_genes.html">three_genes</a></code>

        </dt>
        <dd>Three genes data set</dd>
      </dl><dl><dt>

          <code><a href="trim.html">trim()</a></code>

        </dt>
        <dd>Trimming data frames using a numeric interval</dd>
      </dl><dl><dt>

          <code><a href="trim_tree.html">trim_tree()</a></code>

        </dt>
        <dd>Trim a tree to remove unused sequences</dd>
      </dl><dl><dt>

          <code><a href="uniform_color_scheme.html">uniform_color_scheme()</a></code>

        </dt>
        <dd>Apply a uniform color scheme to dna_segs and comparisons</dd>
      </dl><dl><dt>

          <code><a href="update_comparisons.html">update_comparisons()</a></code>

        </dt>
        <dd>Update comparisons using dna_segs</dd>
      </dl><dl><dt>

          <code><a href="update_dna_segs.html">update_dna_segs()</a></code>

        </dt>
        <dd>Update dna_segs using comparisons</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mike Puijk.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

