<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Creating dna_segs from files — read_dna_seg • genoPlotR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Creating dna_segs from files — read_dna_seg"><meta name="description" content="Functions to parse dna_seg objects from different file formats. Support
is included for the following file formats: GenBank, EMBL, ptt, FASTA, and
tabular data."><meta property="og:description" content="Functions to parse dna_seg objects from different file formats. Support
is included for the following file formats: GenBank, EMBL, ptt, FASTA, and
tabular data."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genoPlotR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/genoPlotR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/wrapper_script.html">Running genoPlotR from a command-line interface</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Creating dna_segs from files</h1>

      <div class="d-none name"><code>read_dna_seg.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Functions to parse <code>dna_seg</code> objects from different file formats. Support
is included for the following file formats: GenBank, EMBL, ptt, FASTA, and
tabular data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_dna_seg_from_file</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  tagsToParse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDS"</span><span class="op">)</span>,</span>
<span>  fileType <span class="op">=</span> <span class="st">"detect"</span>,</span>
<span>  meta_lines <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  gene_type <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  extra_fields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  boundariesToParse <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  read_sequence <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_dna_seg_from_files</span><span class="op">(</span></span>
<span>  <span class="va">files</span>,</span>
<span>  tagsToParse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDS"</span><span class="op">)</span>,</span>
<span>  fileType <span class="op">=</span> <span class="st">"detect"</span>,</span>
<span>  meta_lines <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  gene_type <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  extra_fields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  boundariesToParse <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  read_sequence <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_dna_seg_from_embl</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  tagsToParse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDS"</span><span class="op">)</span>,</span>
<span>  boundariesToParse <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extra_fields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  read_sequence <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  gene_type <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_dna_seg_from_genbank</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  tagsToParse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDS"</span><span class="op">)</span>,</span>
<span>  boundariesToParse <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extra_fields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  read_sequence <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  gene_type <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_dna_seg_from_fasta</span><span class="op">(</span><span class="va">file</span>, read_sequence <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_dna_seg_from_ptt</span><span class="op">(</span><span class="va">file</span>, meta_lines <span class="op">=</span> <span class="fl">2</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_dna_seg_from_tab</span><span class="op">(</span><span class="va">file</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>A character string containing a file path, or a file connection.</p></dd>


<dt id="arg-tagstoparse">tagsToParse<a class="anchor" aria-label="anchor" href="#arg-tagstoparse"></a></dt>
<dd><p>A character vector of tags to parse for GenBank or EMBL
files. Common examples include <code>"CDS"</code>, <code>"gene"</code>, <code>"tRNA"</code>,
<code>"repeat_region"</code>, and <code>"misc_feature"</code>.</p></dd>


<dt id="arg-filetype">fileType<a class="anchor" aria-label="anchor" href="#arg-filetype"></a></dt>
<dd><p>A character string containing the file format to parse.
Must be one of: <code>"genbank"</code>, <code>"embl"</code>, <code>"ptt"</code>, <code>"fasta"</code>, or <code>"detect"</code>. If
<code>"detect"</code> is chosen, then it will attempt to determine the file format
automatically.</p></dd>


<dt id="arg-meta-lines">meta_lines<a class="anchor" aria-label="anchor" href="#arg-meta-lines"></a></dt>
<dd><p>The number of lines in the ptt file that represent
metadata, not counting the header lines. Standard for NCBI files is 2 (name
and length, number of proteins).</p></dd>


<dt id="arg-gene-type">gene_type<a class="anchor" aria-label="anchor" href="#arg-gene-type"></a></dt>
<dd><p>A character string, determines how genes are visualized.
Must be a valid gene type (see <a href="gene_types.html">gene_types</a>). For GenBank and EMBL files, if
this argument is <code>"auto"</code>, the genes will appear as arrows if there are no
introns, and as exons (blocks) when there are introns present.</p></dd>


<dt id="arg-header">header<a class="anchor" aria-label="anchor" href="#arg-header"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, parses the first line of the tabular file
as a header containing column names.</p></dd>


<dt id="arg-extra-fields">extra_fields<a class="anchor" aria-label="anchor" href="#arg-extra-fields"></a></dt>
<dd><p>A character vector of extra fields to parse for GenBank
or EMBL files. These fields will be added as columns in the resulting
<code>dna_seg</code> object, unless the field was always empty.</p></dd>


<dt id="arg-boundariestoparse">boundariesToParse<a class="anchor" aria-label="anchor" href="#arg-boundariestoparse"></a></dt>
<dd><p>A character vector of tags to parse as sequence
boundaries for GenBank or EMBL files. Common examples include <code>"source"</code>,
<code>"contig"</code>, <code>"chromosome"</code>, and <code>"scaffold"</code>.</p></dd>


<dt id="arg-read-sequence">read_sequence<a class="anchor" aria-label="anchor" href="#arg-read-sequence"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, will add a sequence column to the
<code>dna_seg</code> containing the DNA or amino acid sequence of the features.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, reports timings whenever it starts parsing
a file.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments to pass to <a href="dna_seg.html">as.dna_seg</a>.</p></dd>


<dt id="arg-files">files<a class="anchor" aria-label="anchor" href="#arg-files"></a></dt>
<dd><p>A list or character vector containing file paths. Supports
wildcard expansion (e.g. *.txt).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of <code>dna_seg</code> objects for <code>read_dna_seg_from_files</code>, and a
single <code>dna_seg</code> object otherwise.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>GenBank and EMBL files are two commonly used file types that often contain a
great variety of information. To properly extract data from these files, the
user has to choose which features to extract. Commonly 'CDS' features are of
interest, but other feature tags such as <code>"gene"</code>, <code>"misc_feature"</code>, or
<code>"tRNA"</code> may be of interest. Should a feature contain an inner "pseudo" tag
indicating this CDS or gene is a pseudo gene, this will be presented as a
<code>"CDS_pseudo"</code> or a <code>"gene_pseudo"</code> feature type respectively in the
resulting table. In these two file types, the following fields are parsed
(in addition to the mandatory name, start, end, and strand): gi
(from db_xref=GI), uniprot_id, gene, locus_id (from locus_tag=),
proteinid, product, color, and region_plot. The sequence itself from CDS tags
can also be read using <code>read_sequence = TRUE</code>. In addition, extra tags can be
parsed with the argument <code>extra_fields</code>. If there is more than one field for
a given name, only the first one is parsed.</p>
<p>Tab files representing DNA segments should have at least the following
columns: name, start, end, and strand. If these column names are not all
present, or if there is no header at all, then these 4 columns are assumed
to be the first 4 columns of the file, in that order. If the tab file does
have headers, then any additional columns can be supplied as needed, like
line width and type, pch and/ or cex. See <a href="dna_seg.html">dna_seg</a> for more information.
An example:</p>
<table class="table table"><tr><td>name</td><td>start</td><td>end</td><td>strand</td><td>fill</td></tr><tr><td>feat1A</td><td>2</td><td>1345</td><td>1</td><td>blue</td></tr><tr><td>feat1B</td><td>1399</td><td>2034</td><td>1</td><td>red</td></tr><tr><td>feat1C</td><td>2101</td><td>2932</td><td>-1</td><td>grey</td></tr><tr><td>feat1D</td><td>2800</td><td>3120</td><td>1</td><td>green</td></tr></table><p>FASTA files are parsed differently, depending on the first sequence header
(defline) found in the file. When parsing FASTA files from UniprotKB,
metadata is parsed fromthe deflines, creating extra columns in the resulting
<code>dna_seg</code> object, including locus_id, gene, and product. Alternatively,
positional information of features will be parsed when using FASTA files from
ensembl when they contain this information in their headers. In all other
cases, each entry in a FASTA file will result in a single feature, all
concatenated behind each other in the resulting <code>dna_seg</code>. Some support is
included for parsing metadata from FASTA files from NCBI.</p>
<p>Ptt (or protein table) files are a tabular format providing information on
each protein of a genome (or plasmid, or virus, etc).</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="dna_seg.html">dna_seg</a>, <a href="gene_types.html">gene_types</a></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Lionel Guy, Jens Roat Kultima, Mike Puijk</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Read DNA segment from tab</span></span></span>
<span class="r-in"><span><span class="va">dna_seg3_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/dna_seg3.tab'</span>, package <span class="op">=</span> <span class="st">'genoPlotR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dna_seg3</span> <span class="op">&lt;-</span> <span class="fu">read_dna_seg_from_tab</span><span class="op">(</span><span class="va">dna_seg3_file</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## From GenBank file</span></span></span>
<span class="r-in"><span><span class="va">bq_genbank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/BG_plasmid.gbk'</span>, package <span class="op">=</span> <span class="st">'genoPlotR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bq</span> <span class="op">&lt;-</span> <span class="fu">read_dna_seg_from_file</span><span class="op">(</span><span class="va">bq_genbank</span>, fileType <span class="op">=</span> <span class="st">"detect"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Parsing extra fields in the GenBank file</span></span></span>
<span class="r-in"><span><span class="va">bq</span> <span class="op">&lt;-</span> <span class="fu">read_dna_seg_from_file</span><span class="op">(</span><span class="va">bq_genbank</span>,</span></span>
<span class="r-in"><span>                             extra_fields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"db_xref"</span>, <span class="st">"transl_table"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bq</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "name"         "start"        "end"          "strand"       "length"      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6] "gi"           "gene"         "locus_id"     "product"      "proteinid"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11] "feature"      "gene_type"    "seq_origin"   "db_xref"      "transl_table"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [16] "region_plot"  "col"          "fill"         "lty"          "lwd"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [21] "pch"          "cex"         </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## From embl file</span></span></span>
<span class="r-in"><span><span class="va">bq_embl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/BG_plasmid.embl'</span>, package <span class="op">=</span> <span class="st">'genoPlotR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bq</span> <span class="op">&lt;-</span> <span class="fu">read_dna_seg_from_embl</span><span class="op">(</span><span class="va">bq_embl</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## From ptt files</span></span></span>
<span class="r-in"><span><span class="va">bq_ptt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/BQ.ptt'</span>, package <span class="op">=</span> <span class="st">'genoPlotR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bq</span> <span class="op">&lt;-</span> <span class="fu">read_dna_seg_from_ptt</span><span class="op">(</span><span class="va">bq_ptt</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mike Puijk.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

