<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create comparisons between dna_segs by reading in a file of groupings — read_orthogroup • genoPlotR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create comparisons between dna_segs by reading in a file of groupings — read_orthogroup"><meta name="description" content="Functions to create comparisons between dna_seg objects by parsing
(ortho)groups from a file."><meta property="og:description" content="Functions to create comparisons between dna_seg objects by parsing
(ortho)groups from a file."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genoPlotR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/genoPlotR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/wrapper_script.html">Running genoPlotR from a command-line interface</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create comparisons between dna_segs by reading in a file of groupings</h1>

      <div class="d-none name"><code>read_orthogroup.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Functions to create <code>comparisons</code> between <code>dna_seg</code> objects by parsing
(ortho)groups from a file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_orthogroup_from_file</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  <span class="va">dna_segs</span>,</span>
<span>  <span class="va">fileType</span>,</span>
<span>  seg_labels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  id_tag <span class="op">=</span> <span class="st">"locus_id"</span>,</span>
<span>  group_name <span class="op">=</span> <span class="st">"orthogroup"</span>,</span>
<span>  alter_dna_segs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_orthogroup_from_orthomcl</span><span class="op">(</span><span class="va">file</span>, <span class="va">dna_segs</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_orthogroup_from_orthofinder</span><span class="op">(</span><span class="va">file</span>, <span class="va">dna_segs</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_orthogroup_from_mmseqs2</span><span class="op">(</span><span class="va">file</span>, <span class="va">dna_segs</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_orthogroup_from_diamond</span><span class="op">(</span><span class="va">file</span>, <span class="va">dna_segs</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>A character string containing a file path.</p></dd>


<dt id="arg-dna-segs">dna_segs<a class="anchor" aria-label="anchor" href="#arg-dna-segs"></a></dt>
<dd><p>A list of <code>dna_seg</code> objects.</p></dd>


<dt id="arg-filetype">fileType<a class="anchor" aria-label="anchor" href="#arg-filetype"></a></dt>
<dd><p>A character string containing the file format to parse.
Must be one of: <code>"orthomcl"</code>, <code>"orthofinder"</code>, <code>"mmseqs2"</code>, or <code>"diamond"</code>.</p></dd>


<dt id="arg-seg-labels">seg_labels<a class="anchor" aria-label="anchor" href="#arg-seg-labels"></a></dt>
<dd><p>Only for use with <code>"orthofinder"</code> format files, a character
string of <code>dna_seg</code> labels. See details.</p></dd>


<dt id="arg-id-tag">id_tag<a class="anchor" aria-label="anchor" href="#arg-id-tag"></a></dt>
<dd><p>A character string with a <code>dna_seg</code> column. The (gene) names
taken from the (ortho)groups file will be matched to the names found in this
<code>dna_seg</code> column.</p></dd>


<dt id="arg-group-name">group_name<a class="anchor" aria-label="anchor" href="#arg-group-name"></a></dt>
<dd><p>A character string containing a column name. This column
will contain the group names found in the (ortho)groups file and will be
added to the <code>comparisons</code>, as well as the <code>dna_segs</code> when <code>alter_dna_segs</code>
is set to <code>TRUE</code>.</p></dd>


<dt id="arg-alter-dna-segs">alter_dna_segs<a class="anchor" aria-label="anchor" href="#arg-alter-dna-segs"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, a group column will be added to
each <code>dna_seg</code> containing the groupings found in the (ortho)groups file.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, will report a warning when the column
specified by <code>group_name</code> is already present in the <code>dna_segs</code> and will
therefore be overwritten. Has no effect unless <code>alter_dna_segs</code> is set to
<code>TRUE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Arguments to pass to fread and <code>read_orthogroup_from_file</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>With <code>alter_dna_segs = TRUE</code>, a list with 2
named elements: <code>dna_segs</code> and <code>comparisons</code>, which are both lists containing
the <code>dna_seg</code> and <code>comparison</code> objects, respectively.</p>
<p>With <code>alter_dna_segs = FALSE</code>, a list of <code>comparison</code> objects.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>read_orthogroup_from_orthomcl</code>, <code>read_orthogroup_from_orthofinder</code>,
<code>read_orthogroup_from_mmseqs2</code>, and <code>read_orthogroup_from_diamond</code> are
all just convenience functions for <code>read_orthogroup_from_file</code>.</p>
<p>This function was created to create a list of <code>comparisons</code> from a list of
<code>dna_segs</code> and a file that contains orthologous groups of genes
(orthogroups). However, it could theoretically be used for any grouping of
(genetic) elements on a genomic track. For instance, a group could represent
an operon, pathway, or general function. This function creates the
<code>comparisons</code> by linking together columns from <code>dna_segs</code> (specified by
<code>id_tag</code>).</p>
<p>Because <code>"orthofinder"</code> format files contain columns representing the
different genomes as input, this function will attempt to
query only the <code>dna_seg</code> whose label matches the column. <code>dna_seg</code> labels
will be determined automatically, but they can also be provided using the
<code>seg_labels</code> argument. If the labels cannot be matched, it will continue
without matching <code>dna_seg</code> names, querying each <code>dna_seg</code> for each column.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="comparison.html">comparison</a>, <a href="read_comparison.html">read_comparison_from_file</a></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Mike Puijk</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Generate data</span></span></span>
<span class="r-in"><span><span class="va">names1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1_FeatA1"</span>, <span class="st">"1_FeatA2"</span>, <span class="st">"1_FeatB"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">names2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2_FeatA"</span>, <span class="st">"2_FeatB"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">names3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"3_FeatA"</span>, <span class="st">"3_FeatB"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">df1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">names1</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">501</span>, <span class="fl">1501</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">400</span>, <span class="fl">900</span>, <span class="fl">2200</span><span class="op">)</span>, strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">df2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">names2</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">501</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">400</span>, <span class="fl">1200</span><span class="op">)</span>, strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">df3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">names3</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">501</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">400</span>, <span class="fl">1200</span><span class="op">)</span>, strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Create list of dna_segs</span></span></span>
<span class="r-in"><span><span class="va">dna_segs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="dna_seg.html">dna_seg</a></span><span class="op">(</span><span class="va">df1</span><span class="op">)</span>, <span class="fu"><a href="dna_seg.html">dna_seg</a></span><span class="op">(</span><span class="va">df2</span><span class="op">)</span>, <span class="fu"><a href="dna_seg.html">dna_seg</a></span><span class="op">(</span><span class="va">df3</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Read feature groups from OrthoFinder (Orthogroups.tsv) format</span></span></span>
<span class="r-in"><span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/OrthoFinder_format.tsv'</span>, package <span class="op">=</span> <span class="st">'genoPlotR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">full_data</span> <span class="op">&lt;-</span> <span class="fu">read_orthogroup_from_file</span><span class="op">(</span>file <span class="op">=</span> <span class="va">file</span>, dna_segs <span class="op">=</span> <span class="va">dna_segs</span>,</span></span>
<span class="r-in"><span>                                       fileType <span class="op">=</span> <span class="st">"orthofinder"</span>, </span></span>
<span class="r-in"><span>                                       id_tag <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in read_orthogroup_from_file(file = file, dna_segs = dna_segs, fileType = "orthofinder",     id_tag = "name"):</span> Could not find file at "".</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Plot data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span>dna_segs <span class="op">=</span> <span class="va">full_data</span><span class="op">$</span><span class="va">dna_segs</span>,</span></span>
<span class="r-in"><span>              comparisons <span class="op">=</span> <span class="va">full_data</span><span class="op">$</span><span class="va">comparisons</span>,</span></span>
<span class="r-in"><span>              global_color_scheme <span class="op">=</span> <span class="st">"uniform"</span>,</span></span>
<span class="r-in"><span>              alpha_comparisons <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'full_data' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Examples of these groups in the different supported formats:</span></span></span>
<span class="r-in"><span><span class="va">OrthoFinder_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/OrthoFinder_format.tsv'</span>,</span></span>
<span class="r-in"><span>                                package <span class="op">=</span> <span class="st">'genoPlotR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">OrthoMCL_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/OrthoMCL_format.txt'</span>,</span></span>
<span class="r-in"><span>                             package <span class="op">=</span> <span class="st">'genoPlotR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">MMSeqs2_or_DIAMOND_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/MMseqs2_DIAMOND_format.tsv'</span>,</span></span>
<span class="r-in"><span>                                       package <span class="op">=</span> <span class="st">'genoPlotR'</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">OrthoFinder_file</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>file("") only supports open = "w+" and open = "w+b": using the former</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">OrthoMCL_file</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GroupA: 1_FeatA1 1_FeatA2 2_FeatA 3_FeatA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GroupB: 1_FeatB 2_FeatB 3_FeatB</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">MMSeqs2_or_DIAMOND_file</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>file("") only supports open = "w+" and open = "w+b": using the former</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mike Puijk.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

