<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting started with genoPlotR • genoPlotR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with genoPlotR">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genoPlotR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/genoPlotR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/wrapper_script.html">Running genoPlotR from a command-line interface</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Getting started with genoPlotR</h1>
            
      

      <div class="d-none name"><code>genoPlotR.Rmd</code></div>
    </div>

    
    
<p>
<br></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <strong>genoPlotR</strong> package is intended to produce
publication-grade graphics of gene and genome maps. With the amazing
speed of data production of new DNA sequencing techniques and the
increase in the number of software available to compare these sequences,
there is a great need to graphically represent these sequences and their
comparisons. A number of packages already exist (Artemis, ACT, mauve),
but none of them produce easily reproducible, publication-grade
graphics. The goal of this package is to fill in that gap.</p>
<p>This vignette provides an introduction to <strong>genoPlotR</strong>,
providing the user with examples of increasing complexity. It is not
meant as a comprehensive guide to all the functions and options of the
package, but rather as a first approach to the package.</p>
<p>To load the library in an R session, type:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mvpuijk.github.io/genoPlotR/">genoPlotR</a></span><span class="op">)</span></span></code></pre></div>
<p>
<br></p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>Loading the simplest dataset, applying a color scheme and some limits
to the plotting area, adding a tree and some annotations. For more
details about that plot, refer to the first of the examples shown
below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">three_genes</span><span class="op">)</span></span>
<span><span class="va">dna_segs</span> <span class="op">&lt;-</span> <span class="va">three_genes</span><span class="op">$</span><span class="va">dna_segs</span></span>
<span><span class="va">comparisons</span> <span class="op">&lt;-</span> <span class="va">three_genes</span><span class="op">$</span><span class="va">comparisons</span></span>
<span><span class="va">comparisons</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gradient_color_scheme.html">apply_color_scheme</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span><span class="op">)</span>, <span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Huey"</span>, <span class="st">"Dewey"</span>, <span class="st">"Louie"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dna_segs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ade4</span><span class="fu">::</span><span class="fu"><a href="https://adeverse.github.io/ade4/reference/newick2phylog.html" class="external-link">newick2phylog</a></span><span class="op">(</span><span class="st">"(((Huey:4.2,Dewey:3.9):3.1,Louie:7.3):1);"</span><span class="op">)</span></span>
<span><span class="va">mid_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/middle.html">middle</a></span><span class="op">(</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">xlims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">Inf</span>, <span class="op">-</span><span class="cn">Inf</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="cn">Inf</span>, <span class="cn">Inf</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1850</span>, <span class="fl">2800</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotation.html">annotation</a></span><span class="op">(</span>x1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mid_pos</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">end</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                    x2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">end</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                    text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"region1"</span><span class="op">)</span>,</span>
<span>                    rot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"black"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now plotting these three segments:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span>dna_segs <span class="op">=</span> <span class="va">dna_segs</span>, comparisons <span class="op">=</span> <span class="va">comparisons</span>,</span>
<span>              annotations <span class="op">=</span> <span class="va">annot</span>, annotation_height <span class="op">=</span> <span class="fl">1.3</span>,</span>
<span>              tree <span class="op">=</span> <span class="va">tree</span>, tree_width <span class="op">=</span> <span class="fl">2</span>,</span>
<span>              xlims <span class="op">=</span> <span class="va">xlims</span>,</span>
<span>              main <span class="op">=</span> <span class="st">"Comparison of Huey, Dewey and Louie"</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/quick_start_plot-1.png" width="672"></p>
<p>
<br></p>
</div>
<div class="section level2">
<h2 id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h2>
<p>A lot of examples and help are available in the main functions,
i.e. the reading functions (the various <code>read_dna_seg_from*</code>
and <code>read_comparison_from*</code> functions) and the main plotting
function, <code><a href="../reference/plot_gene_map.html">plot_gene_map()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/help.html" class="external-link">help</a></span><span class="op">(</span><span class="st">"read_dna_seg"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/help.html" class="external-link">help</a></span><span class="op">(</span><span class="st">"read_comparison"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/help.html" class="external-link">help</a></span><span class="op">(</span><span class="st">"plot_gene_map"</span><span class="op">)</span></span></code></pre></div>
<p>Additionally, the web pages on <a href="http://genoplotr.r-forge.r-project.org/" class="external-link">R-forge</a> and <a href="https://github.com/MVPuijk/genoPlotR" class="external-link">GitHub</a> provide ways to
get in touch with the <strong>genoPlotR</strong> community and to submit
bugs and feature requests.</p>
<p>
<br></p>
</div>
<div class="section level2">
<h2 id="objects-in-genoplotr">Objects in <strong>genoPlotR</strong><a class="anchor" aria-label="anchor" href="#objects-in-genoplotr"></a>
</h2>
<p>This section will give an overview of the different types of
<strong>R</strong> objects in <strong>genoPlotR</strong>.</p>
<div class="section level3">
<h3 id="dna_seg">dna_seg<a class="anchor" aria-label="anchor" href="#dna_seg"></a>
</h3>
<p>A <code>dna_seg</code> object is a collection of genes or elements
along a genome, to be represented on a map.</p>
<p><code>dna_seg</code> objects need to have 4 columns,
<code>name</code>, <code>start</code>, <code>end</code> and
<code>strand</code>. Extra columns with the names <code>col</code>,
<code>fill</code>, <code>lty</code>, <code>lwd</code>, <code>pch</code>,
<code>cex</code>, <code>gene_type</code>, and <code>region_type</code>
will be used in the plotting process. Other extra columns will be kept
in the object, but are not used.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">names1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feat1"</span>, <span class="st">"feat2"</span>, <span class="st">"feat3"</span><span class="op">)</span></span>
<span><span class="va">starts1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1000</span>, <span class="fl">1050</span><span class="op">)</span></span>
<span><span class="va">ends1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">600</span>, <span class="fl">800</span>, <span class="fl">1345</span><span class="op">)</span></span>
<span><span class="va">strands1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cols1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"grey"</span>, <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="va">df1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">names1</span>, start <span class="op">=</span> <span class="va">starts1</span>, end <span class="op">=</span> <span class="va">ends1</span>,</span>
<span>                  strand <span class="op">=</span> <span class="va">strands1</span>, col <span class="op">=</span> <span class="va">cols1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dna_seg1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dna_seg.html">dna_seg</a></span><span class="op">(</span><span class="va">df1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">dna_seg1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Classes 'dna_seg', 'data.table' and 'data.frame':    3 obs. of  12 variables:</span></span>
<span><span class="co">#&gt;  $ name       : chr  "feat1" "feat2" "feat3"</span></span>
<span><span class="co">#&gt;  $ start      : num  2 1000 1050</span></span>
<span><span class="co">#&gt;  $ end        : num  600 800 1345</span></span>
<span><span class="co">#&gt;  $ strand     : num  -1 -1 1</span></span>
<span><span class="co">#&gt;  $ col        : chr  "blue" "grey" "red"</span></span>
<span><span class="co">#&gt;  $ gene_type  : chr  "arrows" "arrows" "arrows"</span></span>
<span><span class="co">#&gt;  $ region_plot: chr  "NA" "NA" "NA"</span></span>
<span><span class="co">#&gt;  $ fill       : chr  "grey80" "grey80" "grey80"</span></span>
<span><span class="co">#&gt;  $ lty        : num  1 1 1</span></span>
<span><span class="co">#&gt;  $ lwd        : num  1 1 1</span></span>
<span><span class="co">#&gt;  $ pch        : num  8 8 8</span></span>
<span><span class="co">#&gt;  $ cex        : num  1 1 1</span></span>
<span><span class="co">#&gt;  - attr(*, ".internal.selfref")=&lt;externalptr&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparison">comparison<a class="anchor" aria-label="anchor" href="#comparison"></a>
</h3>
<p>A <code>comparison</code> is a collection of similarities,
representing the comparison between two DNA segments.</p>
<p>Objects (<code>data.frame</code>, <code>data.table</code>, or
<code>list</code>) should have at least named elements
<code>start1</code>, <code>end1</code>, <code>start2</code> and
<code>end2</code>. In addition, an optional <code>col</code> column can
be added, that will give the color of each comparison. Additional
numeric columns can be used for automatic color-coding (via
<code><a href="../reference/gradient_color_scheme.html">gradient_color_scheme()</a></code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">starts1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1000</span>, <span class="fl">1050</span><span class="op">)</span></span>
<span><span class="va">ends1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">600</span>, <span class="fl">800</span>, <span class="fl">1345</span><span class="op">)</span></span>
<span><span class="va">starts2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">800</span>, <span class="fl">1200</span><span class="op">)</span></span>
<span><span class="va">ends2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">900</span>, <span class="fl">1100</span>, <span class="fl">1322</span><span class="op">)</span></span>
<span><span class="va">comparison1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparison.html">as.comparison</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>start1 <span class="op">=</span> <span class="va">starts1</span>, end1 <span class="op">=</span> <span class="va">ends1</span>,</span>
<span>                                        start2 <span class="op">=</span> <span class="va">starts2</span>, end2 <span class="op">=</span> <span class="va">ends2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">comparison1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Classes 'comparison', 'data.table' and 'data.frame': 3 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;  $ start1   : num  2 1000 1050</span></span>
<span><span class="co">#&gt;  $ end1     : num  600 800 1345</span></span>
<span><span class="co">#&gt;  $ start2   : num  50 800 1200</span></span>
<span><span class="co">#&gt;  $ end2     : num  900 1100 1322</span></span>
<span><span class="co">#&gt;  $ direction: num  1 -1 1</span></span>
<span><span class="co">#&gt;  - attr(*, ".internal.selfref")=&lt;externalptr&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="annotation">annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h3>
<p>An <code>annotation</code> object is used to annotate a DNA segment.
It has labels attached to positions. Each label can be attached to a
single position or to a range.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mid_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/middle.html">middle</a></span><span class="op">(</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">annot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotation.html">annotation</a></span><span class="op">(</span>x1 <span class="op">=</span> <span class="va">mid_pos</span>, text <span class="op">=</span> <span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">name</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">annot1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Classes 'annotation', 'data.table' and 'data.frame': 3 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;  $ x1   : num  301 900 1198</span></span>
<span><span class="co">#&gt;  $ x2   : logi  NA NA NA</span></span>
<span><span class="co">#&gt;  $ text : chr  "feat1" "feat2" "feat3"</span></span>
<span><span class="co">#&gt;  $ color: chr  "black" "black" "black"</span></span>
<span><span class="co">#&gt;  $ rot  : num  0 0 0</span></span>
<span><span class="co">#&gt;  - attr(*, ".internal.selfref")=&lt;externalptr&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="tree">tree<a class="anchor" aria-label="anchor" href="#tree"></a>
</h3>
<p>A tree description in Newick format can be parsed using the
<code>ade4</code> package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ade4</span><span class="fu">::</span><span class="fu"><a href="https://adeverse.github.io/ade4/reference/newick2phylog.html" class="external-link">newick2phylog</a></span><span class="op">(</span><span class="st">"(((A_aaa:4.2,B_bbb:3.9):3.1,C_ccc:7.3):1);"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">tree</span><span class="op">$</span><span class="va">leaves</span><span class="op">)</span></span>
<span><span class="co">#&gt;  Named num [1:3] 4.2 3.9 7.3</span></span>
<span><span class="co">#&gt;  - attr(*, "names")= chr [1:3] "A_aaa" "B_bbb" "C_ccc"</span></span></code></pre></div>
<p>
<br></p>
</div>
</div>
<div class="section level2">
<h2 id="reading-data">Reading data<a class="anchor" aria-label="anchor" href="#reading-data"></a>
</h2>
<div class="section level3">
<h3 id="dna-segments">DNA segments<a class="anchor" aria-label="anchor" href="#dna-segments"></a>
</h3>
<p>Several formats can be read by <strong>genoPlotR</strong> to produce
<code>dna_seg</code> objects:</p>
<ul>
<li>GenBank files (<code><a href="../reference/read_dna_seg.html">read_dna_seg_from_genbank()</a></code>)</li>
<li>EMBL files (<code><a href="../reference/read_dna_seg.html">read_dna_seg_from_embl()</a></code>)</li>
<li>FASTA files (<code><a href="../reference/read_dna_seg.html">read_dna_seg_from_fasta()</a></code>)</li>
<li>PTT (protein table) files, tabular versions of GenBank files
(<code><a href="../reference/read_dna_seg.html">read_dna_seg_from_ptt()</a></code>)</li>
<li>User generated tabular files
(<code><a href="../reference/read_dna_seg.html">read_dna_seg_from_tab()</a></code>)</li>
</ul>
<p>The function <code><a href="../reference/read_dna_seg.html">read_dna_seg_from_file()</a></code> is a wrapper
function, that will attempt to guess the correct format of the file.</p>
<p>The first three files are common biological formats and can be
downloaded from major databases, such as the <a href="https://www.ncbi.nlm.nih.gov/" class="external-link">NCBI</a> and the <a href="https://www.ebi.ac.uk/ena/browser/home" class="external-link">ENA</a>. The definition of
EMBL and GenBank files can be found <a href="https://ftp.ebi.ac.uk/pub/databases/embl/doc/FT_current.html" class="external-link">here</a>.</p>
</div>
<div class="section level3">
<h3 id="comparisons">Comparisons<a class="anchor" aria-label="anchor" href="#comparisons"></a>
</h3>
<p><strong>genoPlotR</strong> can read tabular files, either
user-generated tab files (<code><a href="../reference/read_comparison.html">read_comparison_from_tab()</a></code>), or
from BLAST output (<code><a href="../reference/read_comparison.html">read_comparison_from_blast()</a></code>). To
produce files that are readable by <strong>genoPlotR</strong>, BLAST+
commands should be run using the <code>outfmt -6</code> or
<code>outfmt -7</code> options.</p>
</div>
<div class="section level3">
<h3 id="mauve-output">Mauve output<a class="anchor" aria-label="anchor" href="#mauve-output"></a>
</h3>
<p>The backbone output of the <a href="https://darlinglab.org/mauve/mauve.html" class="external-link">Mauve genome aligner</a>
can be parsed using <code><a href="../reference/read_mauve_backbone.html">read_mauve_backbone()</a></code> (Tested with
Mauve 2.3.1).</p>
<p>The function will return a list consisting of a list of
<code>dna_seg</code> objects and the corresponding
<code>comparisons</code>.</p>
<p>
<br></p>
</div>
</div>
<div class="section level2">
<h2 id="plotting-data">Plotting data<a class="anchor" aria-label="anchor" href="#plotting-data"></a>
</h2>
<p>There is only one plotting function in <strong>genoPlotR</strong>,
<code><a href="../reference/plot_gene_map.html">plot_gene_map()</a></code>. Many arguments are available, but here is
a list of the most important ones. Check the documentation for a more
thorough description.</p>
<div style="display: grid; display: grid; grid-template-columns: 1fr 9fr; grid-column-gap: 10px;">
<div>
<p><strong><code>dna_segs</code></strong></p>
</div>
<div>
<p>A list of DNA segment objects.</p>
</div>
<div>
<p><strong><code>comparisons</code></strong></p>
</div>
<div>
<p>A list of comparisons. Should contain one element less than
<code>dna_segs</code>.</p>
</div>
<div>
<p><strong><code>tree</code></strong></p>
</div>
<div>
<p>A phylogenetic tree to be plotted at the left of the figure.</p>
</div>
<div>
<p><strong><code>annotations</code></strong></p>
</div>
<div>
<p>An annotation object, or a list of annotations. Will display
annotations to the first, or to all DNA segments, respectively.</p>
</div>
<div>
<p><strong><code>xlims</code></strong></p>
</div>
<div>
<p>A list of even-numbered numeric vectors, giving the borders of
sub-segments to be plotted. The vector <code>c(1,5000,8000,6000)</code>
will display two sub-segments (1 to 5000 and 6000 to 8000), the second
being in reverse orientation.</p>
</div>
<div>
<p><strong><code>main</code></strong></p>
</div>
<div>
<p>A title to the plot.</p>
</div>
<div>
<p><strong><code>scale</code></strong></p>
</div>
<div>
<p>Should a scale be displayed at the bottom right of the plot?</p>
</div>
<div>
<p><strong><code>dna_seg_scale</code></strong></p>
</div>
<div>
<p>Allows to control the addition of scales to each segments. If simply
<code>TRUE</code>, will display a scale on each segment. If a vector, a
scale will be displayed for the corresponding <code>TRUE</code>
element.</p>
</div>
<div>
<p><strong><code>global_color_scheme</code></strong></p>
</div>
<div>
<p>Generate a new set of colors for the DNA segments and/ or
comparisons.</p>
</div>
<div>
<p><strong><code>plot_new</code></strong></p>
</div>
<div>
<p>Turn off to avoid creating a new plot. Especially useful to integrate
a <strong>genoPlotR</strong> plot in a larger figure.</p>
</div>
</div>
<p>
<br></p>
</div>
<div class="section level2">
<h2 id="other-useful-functions">Other useful functions<a class="anchor" aria-label="anchor" href="#other-useful-functions"></a>
</h2>
<p><code><a href="../reference/gradient_color_scheme.html">gradient_color_scheme()</a></code> Makes a set of colors, either a
gray scale or shades of red and blue, that can then be applied to
<code>comparisons</code>.<br><code><a href="../reference/uniform_color_scheme.html">uniform_color_scheme()</a></code> Applies a color scheme to
<code>dna_segs</code> and/ or <code>comparisons</code>.<br><code><a href="../reference/middle.html">middle()</a></code> Useful to get the middle of a gene, especially to
create annotations.<br><code><a href="../reference/auto_annotate.html">auto_annotate()</a></code> Automatically generates annotations from
<code>dna_segs</code>.<br><code><a href="../reference/edit_dna_segs.html">edit_dna_segs()</a></code> Uses a data frame to edit a list of
<code>dna_segs</code>. <code><a href="https://rdrr.io/r/utils/data.html" class="external-link">data()</a></code> Type
<code>data(package = "genoPlotR")</code> to get the full list of
datasets.</p>
<p>
<br></p>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<p>This section gives step-by-step examples, of gradual complexity. The
last one shows how to combine several plots and to annotate already
generated plots.</p>
<p>For all examples, the first step is to load the library.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mvpuijk.github.io/genoPlotR/">genoPlotR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="example-1-a-very-simple-example">Example 1: A very simple example<a class="anchor" aria-label="anchor" href="#example-1-a-very-simple-example"></a>
</h3>
<p>The data used in this example is a dataset included in
<strong>genoPlotR</strong>, but for the sake of demonstration, it will
be recreated from scratch. To retrieve this dataset from the package,
use <code>data(three_genes)</code>.</p>
<div class="section level4">
<h4 id="data-generation-and-a-very-simple-plot">Data generation and a very simple plot<a class="anchor" aria-label="anchor" href="#data-generation-and-a-very-simple-plot"></a>
</h4>
<p>First, three <code>dna_segs</code> are generated, from data frames
with columns <code>name</code>, <code>start</code>, <code>end</code>,
<code>strand</code> and <code>col</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feat1"</span>, <span class="st">"feat2"</span>, <span class="st">"feat3"</span><span class="op">)</span>,</span>
<span>                  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1000</span>, <span class="fl">1050</span><span class="op">)</span>,</span>
<span>                  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">600</span>, <span class="fl">800</span>, <span class="fl">1345</span><span class="op">)</span>,</span>
<span>                  strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"grey"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dna_seg1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dna_seg.html">dna_seg</a></span><span class="op">(</span><span class="va">df1</span><span class="op">)</span></span>
<span><span class="va">df2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feat1"</span>, <span class="st">"feat2"</span>, <span class="st">"feat3"</span><span class="op">)</span>,</span>
<span>                  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">800</span>, <span class="fl">1200</span><span class="op">)</span>,</span>
<span>                  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">900</span>, <span class="fl">1100</span>, <span class="fl">1322</span><span class="op">)</span>,</span>
<span>                  strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"grey"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dna_seg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dna_seg.html">dna_seg</a></span><span class="op">(</span><span class="va">df2</span><span class="op">)</span></span>
<span><span class="va">df3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feat1"</span>, <span class="st">"feat2"</span>, <span class="st">"feat3"</span><span class="op">)</span>,</span>
<span>                  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1899</span>, <span class="fl">2108</span>, <span class="fl">2803</span><span class="op">)</span>,</span>
<span>                  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2034</span>, <span class="fl">2732</span>, <span class="fl">3620</span><span class="op">)</span>,</span>
<span>                  strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dna_seg3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dna_seg.html">dna_seg</a></span><span class="op">(</span><span class="va">df3</span><span class="op">)</span></span>
<span><span class="va">dna_segs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">dna_seg1</span>, <span class="va">dna_seg2</span>, <span class="va">dna_seg3</span><span class="op">)</span></span></code></pre></div>
<p>Then, create two <code>comparison</code> objects from data frames
with columns <code>start1</code>, <code>end1</code>,
<code>start2</code>, <code>end2</code> and <code>col</code>. In the
first comparison, the starts and ends correspond to the genes created
above.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>start1 <span class="op">=</span> <span class="va">dna_seg1</span><span class="op">$</span><span class="va">start</span>,</span>
<span>                  end1 <span class="op">=</span> <span class="va">dna_seg1</span><span class="op">$</span><span class="va">end</span>,</span>
<span>                  start2 <span class="op">=</span> <span class="va">dna_seg2</span><span class="op">$</span><span class="va">start</span>,</span>
<span>                  end2 <span class="op">=</span> <span class="va">dna_seg2</span><span class="op">$</span><span class="va">end</span><span class="op">)</span></span>
<span><span class="va">comparison1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparison.html">comparison</a></span><span class="op">(</span><span class="va">df4</span><span class="op">)</span></span>
<span><span class="va">df5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>start1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">800</span><span class="op">)</span>,</span>
<span>                  end1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">1100</span><span class="op">)</span>,</span>
<span>                  start2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1899</span>, <span class="fl">2732</span><span class="op">)</span>,</span>
<span>                  end2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2034</span>, <span class="fl">2508</span><span class="op">)</span>,</span>
<span>                  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#67000D"</span>, <span class="st">"#08306B"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">comparison2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparison.html">comparison</a></span><span class="op">(</span><span class="va">df5</span><span class="op">)</span></span>
<span><span class="va">comparisons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">comparison1</span>, <span class="va">comparison2</span><span class="op">)</span></span></code></pre></div>
<p>This constitutes all the material required to create a basic
plot.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span>dna_segs <span class="op">=</span> <span class="va">dna_segs</span>, comparisons <span class="op">=</span> <span class="va">comparisons</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex1_basic_plot-1.png" width="576"></p>
</div>
<div class="section level4">
<h4 id="a-more-elaborate-plot">A more elaborate plot<a class="anchor" aria-label="anchor" href="#a-more-elaborate-plot"></a>
</h4>
<p>Many options can be added to this simple plot. To start, the first
comparison will be colored in grey scale, using an arbitrary scale
passed to the function <code><a href="../reference/gradient_color_scheme.html">gradient_color_scheme()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparisons</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gradient_color_scheme.html">gradient_color_scheme</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span><span class="op">)</span>, <span class="st">"grey"</span><span class="op">)</span></span></code></pre></div>
<p>Second, names and a phylogenetic tree are prepared. The function
<code>newick2phylog()</code> from the package <code>ade4</code> is used
to prepare an object suitable for <strong>genoPlotR</strong>. The
elements of the list of <code>dna_segs</code> must be named and
correspond to the labels of the tree object, to avoid confusion.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Huey"</span>, <span class="st">"Dewey"</span>, <span class="st">"Louie"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dna_segs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names</span></span>
<span><span class="va">tree_HDL</span> <span class="op">&lt;-</span> <span class="fu">ade4</span><span class="fu">::</span><span class="fu"><a href="https://adeverse.github.io/ade4/reference/newick2phylog.html" class="external-link">newick2phylog</a></span><span class="op">(</span><span class="st">"(((Huey:4.2,Dewey:3.9):3.1,Louie:7.3):1);"</span><span class="op">)</span></span></code></pre></div>
<p>Third, annotations to the first segment are added. The first gene is
annotated in its middle (thus the use of the <code><a href="../reference/middle.html">middle()</a></code>
function), while a region comprising the second and third gene are
annotated with a square bracket. Note the use of a numeric value for
<code>x1</code> and <code>NA</code> for the first gene, and the use of
two numeric values for the second region.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mid_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/middle.html">middle</a></span><span class="op">(</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotation.html">annotation</a></span><span class="op">(</span>x1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mid_pos</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">end</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                    x2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">end</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                    text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"region1"</span><span class="op">)</span>,</span>
<span>                    rot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"black"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Finally, the plot is drawn, adding a title, allowing 2 inches for the
tree width, and changing the annotation space to 1.3 lines. The
different options can be checked by removing/adding the different
arguments to <code><a href="../reference/plot_gene_map.html">plot_gene_map()</a></code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span>dna_segs <span class="op">=</span> <span class="va">dna_segs</span>, comparisons <span class="op">=</span> <span class="va">comparisons</span>,</span>
<span>              annotations <span class="op">=</span> <span class="va">annot</span>, annotation_height <span class="op">=</span> <span class="fl">1.3</span>,</span>
<span>              tree <span class="op">=</span> <span class="va">tree_HDL</span>, tree_width <span class="op">=</span> <span class="fl">2</span>,</span>
<span>              main <span class="op">=</span> <span class="st">"Comparison of Huey, Dewey and Louie"</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex1_adv_plot-1.png" width="576"></p>
<p>To explore the different options, new windows can be opened by using
<code><a href="https://rdrr.io/r/grDevices/x11.html" class="external-link">x11()</a></code> or <code><a href="https://rdrr.io/r/stats/window.html" class="external-link">window()</a></code>, depending on the OS, or
saved to files using for example <code><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png()</a></code> or
<code><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf()</a></code>. To finish the plot, the function
<code><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off()</a></code> should be called.</p>
<p>The plot can be saved to a file by first adding a call to one of the
graphical device functions of <strong>R</strong>
(e.g. <code><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf()</a></code>, <code><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png()</a></code>, <code><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">jpeg()</a></code>), then
making the call(s) to <code><a href="../reference/plot_gene_map.html">plot_gene_map()</a></code>, and finally the
function <code><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off()</a></code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="example-2-generating-data-online">Example 2: Generating data online<a class="anchor" aria-label="anchor" href="#example-2-generating-data-online"></a>
</h3>
<p>This section will give an example of how data can be retrieved and
generated online, using the resources available at the <a href="http://www.ncbi.nlm.nih.gov/" class="external-link">NCBI</a>. This example was devised
on 5/3/2010, and given the relatively rapid rate of changes in public
databases, it is possible that part or all of this example becomes
obsolete. Please contact the author of this document (<a href="mailto:lionel.guy@ebc.uu.se" class="email">lionel.guy@ebc.uu.se</a>) in such a case.</p>
<p>In this example, the genomes of <em>Bartonella henselae</em> (RefSeq
accession NC_005956) and <em>Bartonella quintana</em> (RefSeq accession
NC_005955) will be compared.</p>
<div class="section level4">
<h4 id="retrieving-dna_seg-data">Retrieving dna_seg data<a class="anchor" aria-label="anchor" href="#retrieving-dna_seg-data"></a>
</h4>
<p>To retrieve GenBank files, the NCBI Nucleotide Database can be
directly queried, provided that the accession number is known. In such a
case, the database can be accessed from the home page of <a href="http://www.ncbi.nlm.nih.gov/" class="external-link">NCBI</a>, by entering the accession
number (e.g. NC_005956) in the search field and by selecting
“Nucleotide” as the database. The GenBank file is then displayed. It can
be saved in a suitable format by clicking on “Send to:”, selecting
“Complete Record” and “File”, and selecting “GenBank (full)” as the
format.</p>
<p>If the accession number is unknown, the <a href="https://www.ncbi.nlm.nih.gov/datasets/genome/" class="external-link">NCBI genome
database</a> can be queried for the name of the organism. From the list
of assemblies, a GenBank format file can be downloaded by selecting
“Download” from the “Action” column, and selecting the “Sequence and
annotation (GBFF)” file type.</p>
<p>Do this for both NC_005956 and NC_005955.</p>
</div>
<div class="section level4">
<h4 id="performing-web-based-blast-and-retrieving-result">Performing web-based BLAST and retrieving result<a class="anchor" aria-label="anchor" href="#performing-web-based-blast-and-retrieving-result"></a>
</h4>
<p>The two previously retrieved genomes can be compared with BLAST,
which can be done locally or online. The online version of BLAST can be
found on the NCBI website <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi" class="external-link">here</a>. From there,
the genomes can be aligned by entering <code>NC_005956</code> as the
query sequence, selecting “Align two or more sequences”, and then
entering <code>NC_005955</code> as the subject sequence. The other
arguments to BLAST can then be modified at will to suit the needs of the
comparison. Once the BLAST search is run, the hit table (which can be
read by <strong>genoPlotR</strong>) can be downloaded by clicking on
“Download” and then “Hit Table(text)” in the result page.</p>
<p>If BLAST+ is installed locally and findable by R, then you can do
this from within R. The function
<code><a href="../reference/comparisons_from_dna_segs.html">comparisons_from_dna_segs()</a></code> makes a list of
<code>comparisons</code> by running BLAST first, and making FASTA files
from GenBank files as necessary. Alternatively, DIAMOND can be chosen as
well using the argument <code>tool = "diamond"</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BH_vs_BQ</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/comparisons_from_dna_segs.html">comparisons_from_dna_segs</a></span><span class="op">(</span>files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NC_005956.gb"</span>, <span class="st">"NC_005955.gb"</span><span class="op">)</span>,</span>
<span>                                       tool <span class="op">=</span> <span class="st">"blast"</span>, algorithm <span class="op">=</span> <span class="st">"blastn"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h4>
<p>Provided that the GenBank files and their comparison have been saved
under the correct names in the directory where R was started, they can
be parsed directly by <strong>genoPlotR</strong>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BH</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_dna_seg.html">read_dna_seg_from_file</a></span><span class="op">(</span><span class="st">"NC_005956.gb"</span><span class="op">)</span></span>
<span><span class="va">BQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_dna_seg.html">read_dna_seg_from_file</a></span><span class="op">(</span><span class="st">"NC_005955.gb"</span><span class="op">)</span></span>
<span><span class="va">BH_vs_BQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_comparison.html">read_comparison_from_blast</a></span><span class="op">(</span><span class="st">"NC_005956_NC_005955"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, <code>data(barto)</code> contains
<code>dna_segs</code> made from these two genomes, as well as a
<code>comparison</code> that was made from BLAST results.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">barto</span><span class="op">)</span></span>
<span><span class="va">BH</span> <span class="op">&lt;-</span> <span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">BQ</span> <span class="op">&lt;-</span> <span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">BH_vs_BQ</span> <span class="op">&lt;-</span> <span class="va">barto</span><span class="op">$</span><span class="va">comparisons</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>The data can now be plotted, using the <code>xlims</code> argument to
plot only the first 50,000 base pairs.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xlims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">50000</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">50000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span>dna_segs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">BH</span>, <span class="va">BQ</span><span class="op">)</span>,</span>
<span>              comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">BH_vs_BQ</span><span class="op">)</span>,</span>
<span>              dna_seg_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BH"</span>, <span class="st">"BQ"</span><span class="op">)</span>,</span>
<span>              xlims <span class="op">=</span> <span class="va">xlims</span>,</span>
<span>              main <span class="op">=</span> <span class="st">"BH vs BQ, comparison of the first 50 kb"</span>,</span>
<span>              gene_type <span class="op">=</span> <span class="st">"side_blocks"</span>,</span>
<span>              alpha_comparisons <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>              global_color_scheme <span class="op">=</span> <span class="st">"gradient"</span>,</span>
<span>              dna_seg_scale <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex2_plot-1.png" width="576"></p>
</div>
</div>
<div class="section level3">
<h3 id="example-3-mauve-alignment-of-four-bartonella-genomes">Example 3: Mauve alignment of four <em>Bartonella</em> genomes<a class="anchor" aria-label="anchor" href="#example-3-mauve-alignment-of-four-bartonella-genomes"></a>
</h3>
<p><strong>genoPlotR</strong> is able to parse the backbone file
produced by Mauve (for details on the format, see <a href="https://darlinglab.org/mauve/user-guide/files.html" class="external-link uri">https://darlinglab.org/mauve/user-guide/files.html</a>).</p>
<p>The elements of the DNA segments are no longer genes, but Mauve
blocks. Similarly, the comparison reflects the correspondences between
these blocks. The strand indicates the orientation of the block with
respect to a reference, which by default is the first genome in the
comparison. A single list is returned, which contains two lists, one
containing the <code>dna_segs</code> and the other one the
<code>comparisons</code>.</p>
<p>The genomes in the backbone file are not named, so it is advised to
be cautious in the order of the names given.</p>
<p>In this example, 4 genomes of <em>Bartonella</em> have been compared
with Mauve 2.3.1. The smaller blocks (smaller than 10 kb) are filtered
out, and the second genome (which is the largest) is taken as the
reference. The dataset can also be accessed in the package by running
<code>data(mauve_bbone)</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbone_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/barto.backbone'</span>, package <span class="op">=</span> <span class="st">'genoPlotR'</span><span class="op">)</span></span>
<span><span class="va">bbone</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_mauve_backbone.html">read_mauve_backbone</a></span><span class="op">(</span><span class="va">bbone_file</span>, ref <span class="op">=</span> <span class="fl">2</span>, filter_low <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span><span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B_bacilliformis"</span>, <span class="st">"B_grahamii"</span>, <span class="st">"B_henselae"</span>, <span class="st">"B_quintana"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bbone</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names</span></span></code></pre></div>
<p>The lengths of the comparisons can be calculated by adding the length
on both sides of the comparisons.</p>
<p>By adding <code>global_color_scheme = "gradient"</code> and
<code>color_scheme_column = "length"</code> Now plotting, using
<code>global_color_scheme()</code> to color the segments according to
their lengths.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">bbone</span><span class="op">$</span><span class="va">comparisons</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">cmp</span> <span class="op">&lt;-</span> <span class="va">bbone</span><span class="op">$</span><span class="va">comparisons</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">bbone</span><span class="op">$</span><span class="va">comparisons</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">length</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">cmp</span><span class="op">$</span><span class="va">end1</span> <span class="op">-</span> <span class="va">cmp</span><span class="op">$</span><span class="va">start1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">cmp</span><span class="op">$</span><span class="va">end2</span> <span class="op">-</span> <span class="va">cmp</span><span class="op">$</span><span class="va">start2</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We can then color the comparisons based on these calculated lengths
using the <code>global_color_scheme</code> and
<code>color_scheme_column</code> arguments.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span>dna_segs <span class="op">=</span> <span class="va">bbone</span><span class="op">$</span><span class="va">dna_segs</span>, </span>
<span>              comparisons <span class="op">=</span> <span class="va">bbone</span><span class="op">$</span><span class="va">comparisons</span>,</span>
<span>              global_color_scheme <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"length"</span>, <span class="st">"increasing"</span>, <span class="st">"red_blue"</span>, <span class="fl">0.7</span><span class="op">)</span>,</span>
<span>              override_color_schemes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex3_plot-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="example-4-several-sub-segments-of-four-bartonella-genomes">Example 4: Several sub-segments of four <em>Bartonella</em>
genomes<a class="anchor" aria-label="anchor" href="#example-4-several-sub-segments-of-four-bartonella-genomes"></a>
</h3>
<p>This examples presents how to use the <code>xlims</code> argument to
represent several sub-segments of the same segment, and how to show them
in a reverse orientation.</p>
<p>The data used here is also a comparison between four
<em>Bartonella</em> genomes (see above), but the comparison has been
performed using BLAST. The dataset is available in the package
(<code>data(barto)</code>).</p>
<p>First, the data is loaded and a tree is created.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">barto</span><span class="op">)</span></span>
<span><span class="va">tree_barto</span> <span class="op">&lt;-</span> <span class="fu">ade4</span><span class="fu">::</span><span class="fu"><a href="https://adeverse.github.io/ade4/reference/newick2phylog.html" class="external-link">newick2phylog</a></span><span class="op">(</span><span class="st">"(BB:2.5,(BG:1.8,(BH:1,BQ:0.8):1.9):3);"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>xlims</code> argument is then created. It is a list with as
many elements as there are <code>dna_segs</code>. Each element is an
even-numbered numeric vector, containing the left and right border of
each sub-segment, consecutively. In the first DNA segment, two
sub-segments are shown: between 1,415,000 and 1,445,000 in a reverse
orientation, and between 1,380,000 and 1,412,000, in “normal”
orientation. In the remaining DNA segments, 3, 3, and 1 subsegments are
shown, respectively, all in normal orientation.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xlims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1445000</span>, <span class="fl">1415000</span>, <span class="fl">1380000</span>, <span class="fl">1412000</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>  <span class="fl">10000</span>,   <span class="fl">45000</span>,   <span class="fl">50000</span>,   <span class="fl">83000</span>, <span class="fl">90000</span>, <span class="fl">120000</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>  <span class="fl">15000</span>,   <span class="fl">36000</span>,   <span class="fl">90000</span>,  <span class="fl">120000</span>, <span class="fl">74000</span>,  <span class="fl">98000</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>   <span class="fl">5000</span>,   <span class="fl">82000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>To complete the example, annotations are added. Genes that have a
gene name (i.e. these for which the name is not the locus or synonym
name) are used to annotate the segment. Only every fourth gene is
annotated, to avoid overlapping the tags. The function
<code><a href="../reference/middle.html">middle()</a></code> is used to retrieve the middle of each element of
the DNA segments.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">mid</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/middle.html">middle</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotation.html">annotation</a></span><span class="op">(</span>x1 <span class="op">=</span> <span class="va">mid</span>, text <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">name</span>, rot <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span>  <span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^[^B]"</span>, <span class="va">annot</span><span class="op">$</span><span class="va">text</span>, perl <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="va">annot</span><span class="op">[</span><span class="va">idx</span><span class="op">[</span><span class="va">idx</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">4</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span>, <span class="op">]</span> </span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Finally, plot the result, using scales on each DNA segment, adding a
title, and not limiting the plotting area to the longest segment, to
allow for a better placement of the sub-segments.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span>, <span class="va">barto</span><span class="op">$</span><span class="va">comparisons</span>, tree <span class="op">=</span> <span class="va">tree_barto</span>,</span>
<span>              annotations <span class="op">=</span> <span class="va">annots</span>,</span>
<span>              xlims <span class="op">=</span> <span class="va">xlims</span>,</span>
<span>              limit_to_longest_dna_seg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              dna_seg_scale <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              main <span class="op">=</span> <span class="st">"Comparison of homologous segments in 4 Bartonella genomes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex4_plot-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="example-5-two-segments-of-the-y-chromsome-in-human-and-chimp">Example 5: Two segments of the Y chromsome in human and chimp<a class="anchor" aria-label="anchor" href="#example-5-two-segments-of-the-y-chromsome-in-human-and-chimp"></a>
</h3>
<p>In this example, the ability to plot introns and exons is
demonstrated.</p>
<p>First, data is loaded. The dataset included in the package is used,
but any GenBank file containing introns and exons can be used. Some
annotations are added: for each segment, the range of each gene is
calculated, and corresponding annotations are created.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">chrY_subseg</span><span class="op">)</span></span>
<span><span class="va">genes_homo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">chrY_subseg</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="va">x_homo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">genes_homo</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>                 <span class="fu"><a href="../reference/range.html">range</a></span><span class="op">(</span><span class="va">chrY_subseg</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>                       <span class="op">[</span><span class="va">chrY_subseg</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span> <span class="op">==</span> <span class="va">x</span>,<span class="op">]</span><span class="op">)</span></span>
<span>                 <span class="op">)</span></span>
<span><span class="va">annot_homo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotation.html">annotation</a></span><span class="op">(</span>x1 <span class="op">=</span> <span class="va">x_homo</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, x2 <span class="op">=</span> <span class="va">x_homo</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>,</span>
<span>                         text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">x_homo</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">genes_pan</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">chrY_subseg</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="va">x_pan</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">genes_pan</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>                <span class="fu"><a href="../reference/range.html">range</a></span><span class="op">(</span><span class="va">chrY_subseg</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span>                      <span class="op">[</span><span class="va">chrY_subseg</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span> <span class="op">==</span> <span class="va">x</span>,<span class="op">]</span><span class="op">)</span></span>
<span>                 <span class="op">)</span></span>
<span><span class="va">annot_pan</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotation.html">annotation</a></span><span class="op">(</span>x1 <span class="op">=</span> <span class="va">x_pan</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, x2 <span class="op">=</span> <span class="va">x_pan</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>,</span>
<span>                        text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">x_pan</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>The segments can be directly plotted, passing the annotations as a
list.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">main</span> <span class="op">&lt;-</span> <span class="st">"Comparison of two subsegments in H. sapiens and P. troglodytes"</span></span>
<span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span><span class="va">chrY_subseg</span><span class="op">$</span><span class="va">dna_segs</span>, <span class="va">chrY_subseg</span><span class="op">$</span><span class="va">comparison</span>,</span>
<span>              annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">annot_homo</span>, <span class="va">annot_pan</span><span class="op">)</span>,</span>
<span>              dna_seg_scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              main <span class="op">=</span> <span class="va">main</span>,</span>
<span>              scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex5_plot-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="example-6-combining-several-genoplotr-figures-and-annotating-the-figure">Example 6: Combining several <strong>genoPlotR</strong> figures and
annotating the figure<a class="anchor" aria-label="anchor" href="#example-6-combining-several-genoplotr-figures-and-annotating-the-figure"></a>
</h3>
<p>In this example, some of the previous plots are combined in a single,
multi-panel figure. Using the tools present in the <code>grid</code>
package, the plot is annotated further. More information is available in
the documentation of the <code>grid</code> package.</p>
<p>The example here uses <strong>R</strong> objects built in the
previous examples, which should thus be run first.</p>
<p>First, a <code>viewport</code> that will contain all plots is pushed
in a new page. The plot is divided in three rows, of relative heights 1,
1.3 and 0.8.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">pushViewport</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grid/viewport.html" class="external-link">viewport</a></span><span class="op">(</span></span>
<span>  layout <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/grid.layout.html" class="external-link">grid.layout</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">1</span>,</span>
<span>                     heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1.3</span>,<span class="fl">0.8</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"null"</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"overall_vp"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The three panels A to C, containing the result of the above examples
4 to 6 are pushed into viewports, calling <code><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">upViewport()</a></code>
after each plot to come back to the main viewport. After the last plot,
<code>upViewport(0)</code> is called to go up 2 viewports, coming back
to the root viewport. Each time, the argument
<code>plot_new = FALSE</code> is used, to avoid plotting in a new page
every single panel.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Panel A</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">pushViewport</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grid/viewport.html" class="external-link">viewport</a></span><span class="op">(</span>layout.pos.row <span class="op">=</span> <span class="fl">1</span>, name <span class="op">=</span> <span class="st">"panelA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span>dna_segs <span class="op">=</span> <span class="va">bbone</span><span class="op">$</span><span class="va">dna_segs</span>,</span>
<span>              comparisons <span class="op">=</span> <span class="va">bbone</span><span class="op">$</span><span class="va">comparisons</span>,</span>
<span>              dna_seg_scale <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>              scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              main <span class="op">=</span> <span class="st">"A"</span>, main_pos <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>              plot_new <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">upViewport</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## Panel B</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">pushViewport</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grid/viewport.html" class="external-link">viewport</a></span><span class="op">(</span>layout.pos.row <span class="op">=</span> <span class="fl">2</span>, name <span class="op">=</span> <span class="st">"panelB"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span>, <span class="va">barto</span><span class="op">$</span><span class="va">comparisons</span>, </span>
<span>              annotations <span class="op">=</span> <span class="va">annots</span>,</span>
<span>              tree <span class="op">=</span> <span class="va">tree_barto</span>,</span>
<span>              xlims <span class="op">=</span> <span class="va">xlims</span>,</span>
<span>              limit_to_longest_dna_seg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              dna_seg_scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              main <span class="op">=</span> <span class="st">"B"</span>, main_pos <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>              annotation_height <span class="op">=</span> <span class="fl">0.6</span>, annotation_cex <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>              plot_new <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">upViewport</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## Panel C</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">pushViewport</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grid/viewport.html" class="external-link">viewport</a></span><span class="op">(</span>layout.pos.row <span class="op">=</span> <span class="fl">3</span>, name <span class="op">=</span> <span class="st">"panelC"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span><span class="va">chrY_subseg</span><span class="op">$</span><span class="va">dna_segs</span>, <span class="va">chrY_subseg</span><span class="op">$</span><span class="va">comparison</span>,</span>
<span>              annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">annot_homo</span>, <span class="va">annot_pan</span><span class="op">)</span>,</span>
<span>              dna_seg_scale <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              main <span class="op">=</span> <span class="st">"C"</span>, main_pos <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>              plot_new <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">upViewport</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex6_plot-1.png" width="672"></p>
<p>The functions <code><a href="https://rdrr.io/r/grid/current.viewport.html" class="external-link">current.vpTree()</a></code> and
<code><a href="https://rdrr.io/r/grid/grid.ls.html" class="external-link">grid.ls()</a></code> can be used to see the current viewport
structure, and to see all object names.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grid_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grid/grid.ls.html" class="external-link">grid.ls</a></span><span class="op">(</span>grob <span class="op">=</span> <span class="cn">TRUE</span>, viewports <span class="op">=</span> <span class="cn">TRUE</span>, print <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">grid_list</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/current.viewport.html" class="external-link">current.vpTree</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; List of 6</span></span>
<span><span class="co">#&gt;  $ name   : chr [1:1334] "ROOT" "overall_vp" "panelA" "oma" ...</span></span>
<span><span class="co">#&gt;  $ gDepth : num [1:1334] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;  $ vpDepth: num [1:1334] 0 1 2 3 4 5 6 6 5 6 ...</span></span>
<span><span class="co">#&gt;  $ gPath  : chr [1:1334] "" "" "" "" ...</span></span>
<span><span class="co">#&gt;  $ vpPath : chr [1:1334] "" "ROOT" "ROOT::overall_vp" "ROOT::overall_vp::panelA" ...</span></span>
<span><span class="co">#&gt;  $ type   : chr [1:1334] "vpListing" "vpListing" "vpListing" "vpListing" ...</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr "flatGridListing"</span></span>
<span><span class="co">#&gt; viewport[ROOT]-&gt;(viewport[overall_vp]-&gt;(viewport[panelA]-&gt;(viewport[oma]-&gt;(viewport[oma_layout]-&gt;(viewport[main], viewport[frame]-&gt;(viewport[tree_outer]-&gt;(viewport[tree]-&gt;(viewport[labels])), viewport[plotarea_outer]-&gt;(viewport[plotarea]-&gt;(viewport[map]-&gt;(viewport[comparison.1], viewport[comparison.2], viewport[comparison.3], viewport[seg_plot.1], viewport[seg_plot.2], viewport[seg_plot.3], viewport[seg_plot.4], viewport[scale_and_dna_seg.1]-&gt;(viewport[dna_seg.1.1]), viewport[scale_and_dna_seg.2]-&gt;(viewport[dna_seg.2.1]), viewport[scale_and_dna_seg.3]-&gt;(viewport[dna_seg.3.1]), viewport[scale_and_dna_seg.4]-&gt;(viewport[dna_seg_scale.4.1], viewport[dna_seg.4.1])))))))), viewport[panelB]-&gt;(viewport[oma]-&gt;(viewport[oma_layout]-&gt;(viewport[main], viewport[frame]-&gt;(viewport[tree_outer]-&gt;(viewport[tree]-&gt;(viewport[treeFrame]-&gt;(viewport[GRID.VP.13]-&gt;(viewport[GRID.VP.14]-&gt;(viewport[tree.branches]), viewport[GRID.VP.15]-&gt;(viewport[tree.labels]))))), viewport[plotarea_outer]-&gt;(viewport[plotarea]-&gt;(viewport[map]-&gt;(viewport[comparison.1], viewport[comparison.2], viewport[comparison.3], viewport[seg_plot.1], viewport[seg_plot.2], viewport[seg_plot.3], viewport[seg_plot.4], viewport[scale_and_dna_seg.1]-&gt;(viewport[dna_seg_scale.1.1], viewport[dna_seg_scale.1.2], viewport[dna_seg.1.1], viewport[annotation.1.1], viewport[gap.1.2], viewport[dna_seg.1.2], viewport[annotation.1.2]), viewport[scale_and_dna_seg.2]-&gt;(viewport[gap.2.2], viewport[gap.2.3], viewport[dna_seg_scale.2.1], viewport[dna_seg_scale.2.2], viewport[dna_seg_scale.2.3], viewport[dna_seg.2.1], viewport[dna_seg.2.2], viewport[dna_seg.2.3], viewport[annotation.2.1], viewport[annotation.2.2], viewport[annotation.2.3]), viewport[scale_and_dna_seg.3]-&gt;(viewport[gap.3.2], viewport[gap.3.3], viewport[dna_seg_scale.3.1], viewport[dna_seg_scale.3.2], viewport[dna_seg_scale.3.3], viewport[dna_seg.3.1], viewport[annotation.3.1], viewport[dna_seg.3.2], viewport[annotation.3.2], viewport[dna_seg.3.3], viewport[annotation.3.3]), viewport[scale_and_dna_seg.4]-&gt;(viewport[annotation.4.1], viewport[dna_seg_scale.4.1], viewport[dna_seg.4.1])))))))), viewport[panelC]-&gt;(viewport[oma]-&gt;(viewport[oma_layout]-&gt;(viewport[main], viewport[frame]-&gt;(viewport[plotarea_outer]-&gt;(viewport[plotarea]-&gt;(viewport[map]-&gt;(viewport[comparison.1], viewport[seg_plot.1], viewport[seg_plot.2], viewport[scale_and_dna_seg.1]-&gt;(viewport[dna_seg_scale.1.1], viewport[dna_seg.1.1], viewport[annotation.1.1]), viewport[scale_and_dna_seg.2]-&gt;(viewport[dna_seg_scale.2.1], viewport[dna_seg.2.1], viewport[annotation.2.1]))))))))))</span></span></code></pre>
<p>Despite the complexity of the viewport structure, it is possible to
identify the previously defined panel A, B and C. Using the
<code><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">downViewport()</a></code> function, it will be possible to find and
modify elements of the plots.</p>
<p>First, the labels of panel A will be modified to have the names in
italics and to replace the underscores by a dot and a space. The second
label is also removed. Note the use of <code><a href="https://rdrr.io/r/grid/grid.edit.html" class="external-link">grid.edit()</a></code> and
<code><a href="https://rdrr.io/r/grid/grid.remove.html" class="external-link">grid.remove()</a></code>.</p>
<p>Second, a red rectangle will be placed around one subsegment of panel
B.</p>
<p>The plot is shown before and after these 2 edits.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">downViewport</a></span><span class="op">(</span><span class="st">"panelA"</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">names</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">new_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">". "</span>, <span class="va">names</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/grid/grid.edit.html" class="external-link">grid.edit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"label"</span>, <span class="va">i</span>, sep <span class="op">=</span> <span class="st">"."</span><span class="op">)</span>,</span>
<span>            label <span class="op">=</span> <span class="va">new_label</span>, </span>
<span>            gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontface <span class="op">=</span> <span class="st">"italic"</span><span class="op">)</span>,</span>
<span>            redraw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/grid.remove.html" class="external-link">grid.remove</a></span><span class="op">(</span><span class="st">"label.2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">upViewport</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">downViewport</a></span><span class="op">(</span><span class="st">"panelB"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">downViewport</a></span><span class="op">(</span><span class="st">"dna_seg.3.2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">grid.rect</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2.2</span>, <span class="st">"npc"</span><span class="op">)</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="st">"red"</span>, lwd <span class="op">=</span> <span class="fl">2</span>, fill <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">upViewport</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex6_mod_panela-1.png" width="672"><img src="genoPlotR_files/figure-html/ex6_mod_panela-2.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="example-7-using-user-defined-gene_type-and-seg_plot">Example 7: Using user-defined gene_type and seg_plot<a class="anchor" aria-label="anchor" href="#example-7-using-user-defined-gene_type-and-seg_plot"></a>
</h3>
<p>In this example, the use of personalized <code>gene_types</code> and
of <code>seg_plots</code> to show data next to DNA segments is
demonstrated. The same data as in example 4 is used.</p>
<p>A graphical function is defined, to represent triangles instead of
arrows for genes. The triangles are oriented in function of their strand
and the transparency is set differently according to a length cut-off,
passed with the function. As required, the function has only two
arguments, a <code>gene</code> which is a row of a <code>dna_seg</code>,
and the dots (<code>...</code>). Any extra argument passed to
<code><a href="../reference/plot_gene_map.html">plot_gene_map()</a></code> will be passed in turn to the
<code>gene_type</code>-defining function.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">starGrob</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">gene</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co">## Coordinates for the star</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2.5</span><span class="op">)</span> <span class="op">*</span> <span class="va">pi</span><span class="op">)</span> <span class="op">*</span> <span class="op">(</span><span class="va">gene</span><span class="op">$</span><span class="va">end</span> <span class="op">-</span> <span class="va">gene</span><span class="op">$</span><span class="va">start</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span> <span class="op">+</span> <span class="op">(</span><span class="va">gene</span><span class="op">$</span><span class="va">end</span> <span class="op">+</span> <span class="va">gene</span><span class="op">$</span><span class="va">start</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span>  <span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2.5</span><span class="op">)</span> <span class="op">*</span> <span class="va">pi</span><span class="op">)</span> <span class="op">*</span> <span class="va">gene</span><span class="op">$</span><span class="va">strand</span> <span class="op">*</span> <span class="fl">0.5</span> <span class="op">+</span> <span class="fl">0.5</span></span>
<span>  <span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="co">## Attribute line_col only if present in the gene</span></span>
<span>  <span class="va">line_col</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">gene</span><span class="op">$</span><span class="va">line_col</span><span class="op">)</span><span class="op">)</span> <span class="va">gene</span><span class="op">$</span><span class="va">line_col</span> <span class="kw">else</span> <span class="va">gene</span><span class="op">$</span><span class="va">col</span></span>
<span>  <span class="co">## Having a conditional transparency, depending on a length cut-off</span></span>
<span>  <span class="co">## passed via dots</span></span>
<span>  <span class="va">length_cutoff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span><span class="op">$</span><span class="va">length_cutoff</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">length_cutoff</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">alpha</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="op">(</span><span class="va">gene</span><span class="op">$</span><span class="va">end</span><span class="op">-</span><span class="va">gene</span><span class="op">$</span><span class="va">start</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">length_cutoff</span><span class="op">)</span> <span class="fl">0.3</span> <span class="kw">else</span> <span class="fl">0.8</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co">## Grobs</span></span>
<span>  <span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grid/grid.polygon.html" class="external-link">polygonGrob</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span>, <span class="va">y</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span>,</span>
<span>                   gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">gene</span><span class="op">$</span><span class="va">col</span>, col <span class="op">=</span> <span class="va">line_col</span>,</span>
<span>                             lty <span class="op">=</span> <span class="va">gene</span><span class="op">$</span><span class="va">lty</span>, lwd <span class="op">=</span> <span class="va">gene</span><span class="op">$</span><span class="va">lwd</span>, alpha <span class="op">=</span> <span class="va">alpha</span><span class="op">)</span>,</span>
<span>                   default.units <span class="op">=</span> <span class="st">"native"</span><span class="op">)</span></span>
<span>  <span class="va">t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grid/grid.text.html" class="external-link">textGrob</a></span><span class="op">(</span>label <span class="op">=</span> <span class="st">"***"</span>,</span>
<span>                x <span class="op">=</span> <span class="op">(</span><span class="va">gene</span><span class="op">$</span><span class="va">end</span><span class="op">+</span><span class="va">gene</span><span class="op">$</span><span class="va">start</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span>,</span>
<span>                y <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                default.units <span class="op">=</span> <span class="st">"native"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/grid/grid.grob.html" class="external-link">gList</a></span><span class="op">(</span><span class="va">g</span>, <span class="va">t</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>To make use of that, the <code>gene_type</code> column of the
<code>dna_segs</code> must be replaced. To make use of the supplementary
argument <code>line_col</code>, a column is added to the
<code>dna_seg</code> object. Finally, plot the result.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene_type</span> <span class="op">&lt;-</span> <span class="st">"starGrob"</span></span>
<span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene_type</span> <span class="op">&lt;-</span> <span class="st">"starGrob"</span></span>
<span><span class="va">line_col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">line_col</span> <span class="op">&lt;-</span> <span class="va">line_col</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span>, <span class="va">barto</span><span class="op">$</span><span class="va">comparisons</span>,</span>
<span>              tree <span class="op">=</span> <span class="va">tree_barto</span>,</span>
<span>              annotations <span class="op">=</span> <span class="va">annots</span>,</span>
<span>              xlims <span class="op">=</span> <span class="va">xlims</span>, </span>
<span>              dna_seg_scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              length_cutoff <span class="op">=</span> <span class="fl">600</span>,</span>
<span>              annotation_height <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex7_replace_geneType-1.png" width="672"></p>
<p>Now, let’s plot some random data along with the DNA segment. A
standard <code>grid</code> function returning a grob is used, but it is
possible to use personalized functions, as long as they return
<code>grob</code> or <code>gList</code> objects. For example, the
function <code>starGrob()</code> defined in the previous exercise could
be modified to be used here.</p>
<p>The <code>seg_plot</code> objects are first defined, with the
following structure:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seg_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ds</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="../reference/range.html">range</a></span><span class="op">(</span><span class="va">ds</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, by <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span>  <span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/jitter.html" class="external-link">jitter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">300</span>, length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span>, amount <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/seg_plot.html">seg_plot</a></span><span class="op">(</span>func <span class="op">=</span> <span class="va">linesGrob</span>, args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>,</span>
<span>                                         gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/gray.html" class="external-link">grey</a></span><span class="op">(</span><span class="fl">0.3</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">seg_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 5</span></span>
<span><span class="co">#&gt;  $ func :function (x = unit(c(0, 1), "npc"), y = unit(c(0, 1), "npc"), default.units = "npc", </span></span>
<span><span class="co">#&gt;     arrow = NULL, name = NULL, gp = gpar(), vp = NULL)  </span></span>
<span><span class="co">#&gt;  $ args :List of 4</span></span>
<span><span class="co">#&gt;   ..$ x            : num [1:1445] 1 1001 2001 3001 4001 ...</span></span>
<span><span class="co">#&gt;   ..$ y            : num [1:1445] 58.1 133.6 110.4 66.1 51.3 ...</span></span>
<span><span class="co">#&gt;   ..$ gp           :List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ col: chr "#4D4D4D"</span></span>
<span><span class="co">#&gt;   .. ..$ lty: num 2</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "class")= chr "gpar"</span></span>
<span><span class="co">#&gt;   ..$ default.units: chr "native"</span></span>
<span><span class="co">#&gt;  $ xargs: chr [1:5] "x" "x0" "x1" "x2" ...</span></span>
<span><span class="co">#&gt;  $ yargs: chr [1:5] "y" "y0" "y1" "y2" ...</span></span>
<span><span class="co">#&gt;  $ ylim : NULL</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr [1:2] "seg_plot" "list"</span></span></code></pre></div>
<p>Each <code>seg_plot</code> element has a <code>func</code> argument,
giving the function to use to plot the data, here the <code>grid</code>
function <code><a href="https://rdrr.io/r/grid/grid.lines.html" class="external-link">linesGrob()</a></code>. The arguments that are passed to this
function are contained in the <code>args</code> argument, a list. Notice
the use of <code>gp = gpar(...)</code>.</p>
<p>The result can then be plotted, and modulated using the arguments
<code>seg_plot_height</code>, <code>seg_plot_height_unit</code>,
<code>seg_plot_yaxis</code>, and <code>seg_plot_yaxis_cex</code>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gene_map.html">plot_gene_map</a></span><span class="op">(</span><span class="va">barto</span><span class="op">$</span><span class="va">dna_segs</span>, <span class="va">barto</span><span class="op">$</span><span class="va">comparisons</span>, tree <span class="op">=</span> <span class="va">tree_barto</span>,</span>
<span>              annotations <span class="op">=</span> <span class="va">annots</span>,</span>
<span>              xlims <span class="op">=</span> <span class="va">xlims</span>,</span>
<span>              seg_plots <span class="op">=</span> <span class="va">seg_plots</span>,</span>
<span>              seg_plot_height <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>              seg_plot_height_unit <span class="op">=</span> <span class="st">"null"</span>,</span>
<span>              seg_plot_yaxis <span class="op">=</span> <span class="fl">2</span>,</span>
<span>              seg_plot_yaxis_cex <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="genoPlotR_files/figure-html/ex7_plotwithsegplots-1.png" width="672"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mike Puijk.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
