<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Running genoPlotR from a command-line interface • genoPlotR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running genoPlotR from a command-line interface">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">genoPlotR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/genoPlotR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/wrapper_script.html">Running genoPlotR from a command-line interface</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Running genoPlotR from a command-line interface</h1>
            
      

      <div class="d-none name"><code>wrapper_script.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="run_genoplotr-r">run_genoPlotR.R<a class="anchor" aria-label="anchor" href="#run_genoplotr-r"></a>
</h2>
<p>The R script <code>run_genoPlotR.R</code> is a wrapper script that
can generate gene maps from a single command.
<code>run_genoPlotR.R</code> was created so that the majority of
features from genoPlotR are accessible directly from a command-line
interface. As such, there are many customization options available.</p>
<p>At minimum, R and the genoPlotR package must be installed to run
<code>run_genoPlotR.R</code>. Some of the features of
<code>run_genoPlotR.R</code> require the command-line versions of BLAST
or DIAMOND to be installed as well.</p>
<div class="section level3">
<h3 id="basic-usage">Basic usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h3>
<p>The most prominent use-case of <code>run_genoPlotR.R</code> is to
create regional plots of multiple genomes at a time. This usually
involves parsing a set of GenBank files and then using BLAST to create
comparisons between the resulting DNA segments.</p>
<p>An example of such a command is shown below:</p>
<pre><code>Rscript run_genoPlotR.R \
--seg_files "path/to/dataset/GenBank/*" \
--comp_path "path/to/dataset/output/comps" \
--comp_format blastp \
--ids "path/to/dataset/ids.tab" \
--out "path/to/dataset/output/genemap.pdf"</code></pre>
<p>The GenBank files are provided by the <code>--seg_files</code>
option, which supports the usage of wildcards (<code>*</code>) to select
multiple files at once. In this case, all files in the
<code>GenBank</code> folder will be selected.</p>
<p>The folder specified by <code>--comp_path</code> holds the files used
to create comparisons between the DNA segments. The
<code>--comp_format</code> option tells <code>run_genoPlotR.R</code>
which format to use for the comparisons. If DIAMOND or a BLAST program
is selected, like <code>blastp</code> here, <code>run_genoPlotR.R</code>
will run that program to generate the necessary files, unless the files
already exist in the provided location.</p>
<p>The <code>--ids</code> option is often <strong>essential</strong>
when you want to make a regional plot. It specifies the <strong>IDs
file</strong>, which can be used to select which features to plot. The
<strong>IDs file</strong> is a table of IDs that correspond to features
to plot in the form of a tab-delimited file. Below is an example of the
contents of an <strong>IDs file</strong>:</p>
<table class="table">
<thead><tr class="header">
<th>id</th>
<th>seg_label</th>
<th>fill</th>
<th>region_plot</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>coaE</td>
<td>Bhenselae</td>
<td>red</td>
<td>true</td>
</tr>
<tr class="even">
<td>coaE</td>
<td>Bquintana</td>
<td>blue</td>
<td>true</td>
</tr>
<tr class="odd">
<td>dnaQ</td>
<td></td>
<td>green</td>
<td></td>
</tr>
<tr class="even">
<td>AYT27_RS00030</td>
<td></td>
<td>#CA5050</td>
<td>true</td>
</tr>
</tbody>
</table>
<p>The <code>fill</code> column changes the color of the selected
features, while adding the value <code>true</code> to the
<code>region_plot</code> column marks the feature to be plotted. In
principle, any column can be included in an <strong>IDs file</strong>.
These columns and the respective values will be added to the
<code>dna_segs</code>, overwriting existing data if those columns were
already present.</p>
<p>However, The <code>id</code> and <code>seg_label</code> columns both
have special meanings and are treated differently:</p>
<ul>
<li>
<code>id</code> is the only mandatory column. It will match the
values found in this column with those from the <code>dna_segs</code>,
by default it will look in the <code>name</code> and
<code>locus_id</code> columns (this can be changed with the
<code>--id_tags</code> option). The IDs do not have to match exactly.
Internally, <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep()</a></code> is used, so wildcards (<code>*</code>)
and regular expressions will work as well.</li>
<li>The <code>seg_label</code> column allows you to specify a single DNA
segment by name. By doing so, you can edit features in only a single
<code>dna_seg</code>, even when the ID in question can be found in
multiple <code>dna_segs</code>. If a row is left empty, or this column
is omitted entirely, then all <code>dna_segs</code> will be queried for
each ID.</li>
</ul>
</div>
<div class="section level3">
<h3 id="options">Options<a class="anchor" aria-label="anchor" href="#options"></a>
</h3>
<div class="section level4">
<h4 id="dna-segments">DNA segments<a class="anchor" aria-label="anchor" href="#dna-segments"></a>
</h4>
<strong><code>-s ["FILE PATH1,FILE PATH2"], --seg_files=["FILE PATH1,FILE PATH2"]</code></strong>
<ul>
(Mandatory): Path to the dna_seg files to read in, with file names
separated by commas. GenBank files are recommended. Each file will
become a single dna_seg, so genomes consisting of multiple contigs or
chromosomes should be concatenated into a single file. Alternatively, a
single file with the suffix <code>.RDS</code> can be provided, which
will be read as an R object containing a list of dna_segs.
</ul>
<strong><code>-i [FILE PATH], --ids=[FILE PATH]</code></strong>
<ul>
<p>A tab-delimited file with IDs corresponding to features that should
be plotted. The header should at least contain an <code>id</code>
column, but can include other column names to add to the dna_seg
objects. The IDs can refer to values from any dna_seg column, specified
by the <code>--id_tags</code> option. IMPORTANT: If
<code>--comp_format</code> is a blastp algorithm or DIAMOND, then the
<code>region_plot</code> column will be carried over to the next dna_seg
object, for any features linked by the comparisons. Notable columns
include:</p>
<ul>
<li>id: An identifying feature ID</li>
<li>seg_label: The name of the dna_seg object to query</li>
<li>fill: Anything identifiable as a color by R (e.g. <code>red</code>,
<code>blue4</code>, <code>salmon</code>, <code>#88CCEE</code>)</li>
<li>region_plot: If the value is <code>plot</code>, <code>true</code>,
<code>yes</code>, or <code>y</code>, then the genome neighbourhood of
that feature will be plotted.</li>
</ul>
</ul>
<strong><code>-I ["COLUMN1,COLUMN2"], --id_tags=["COLUMN1,COLUMN2"]</code></strong>
<ul>
A list of dna_seg column names to use alongside <code>--ids</code>. The
IDs provided by the <code>--ids</code> option will be used as queries in
the columns provided by this option. Column names must be separated by
commas. (default: <code>"name, locus_id"</code>)
</ul>
<strong><code>--dna_seg_mode=[DNA_SEG PARSING MODE]</code></strong>
<ul>
How should the dna_seg files supplied by <code>--seg_files</code> be
read? Must be one of: <code>all</code>, <code>ids</code>,
<code>fast</code> - all: Read all dna_seg files provided. - ids: Read
only the dna_seg files mentioned in the <code>--ids</code> file, using
the <code>seg_label</code> column. - fast: Only allowed when making a
regional plot using blastp or DIAMOND. After reading in a dna_seg, an
alignment is made between it and the next dna_seg, transferring over the
<code>region_plot</code> column. If there are no features to be plotted
in this next dna_seg, then it is skipped.
</ul>
<strong><code>-c [COMPARISON FORMAT], --comp_format=[COMPARISON FORMAT]</code></strong>
<ul>
<p>The format to use for reading in comparison files. If the necessary
files are not found in <code>--comp_path</code>, then BLAST or DIAMOND
can be chosen here to create them. Must be one of: <code>diamond</code>,
<code>blast</code>, <code>tab</code>, <code>orthofinder</code>,
<code>orthomcl,</code>mmseqs2<code>, or one of the supported blast         programs (</code>blastp<code>,</code>blastp-fast<code>,</code>blastp-short<code>,</code>blastn<code>,</code>blastn-short<code>,</code>megablast<code>,</code>dc-megablast`).</p>
<ul>
<li>diamond: tab-separated DIAMOND result file
(<code>--outfmt 6</code>)</li>
<li>blast: tab-separated BLAST result file (<code>-outfmt 6</code>)</li>
<li>tab: a custom tab-delimited file with the following structure:
<code>"start1 end1 start2 end2 col gene1 gene2"</code>
</li>
<li>orthofinder: An OrthoFinder format file containing orthogroups for
the provided dna_segs</li>
<li>orthomcl: A file containing orthogroups in the orthomcl format</li>
<li>mmseqs2: A file containing clusters of genes in the “Cluster TSV
format” from mmseqs2, or the cluster output from DIAMOND</li>
</ul>
</ul>
<strong><code>-C [DIRECTORY/FILE PATH], --comp_path=[DIRECTORY/FILE PATH]</code></strong>
<ul>
Path to the directory for the comparison files, or the path to a file
containing orthogroups, if <code>--comp_format</code> is
<code>orthofinder</code> or <code>orthomcl</code>. This option is
ignored if <code>--comp_format</code> is not provided. With a
<code>--comp_format</code> of DIAMOND or any BLAST program, this script
will attempt to create the necessary comparison files if it can not
locate them in the directory provided. The file names should contain the
names of the dna_segs joined by an underscore,
e.g. <code>"Genome1_Genome2.suffix"</code> default: current directory
</ul>
<strong><code>--comp_mode=[COMPARISON MODE]</code></strong>
<ul>
The filter mode used when sequentially parsing the sequence alignments
for each dna_seg object. This option is ignored unless a BLAST program
or DIAMOND was chosen using <code>--comp_format</code>. It is
recommended to use <code>besthit</code> and <code>bidirectional</code>
only when doing comparisons on a list of features, like genes or
proteins. Must be one of: <code>besthit</code>,
<code>bidirectional</code>, <code>full</code>. (default:
<code>full</code>) - besthit: Filters the sequence alignment results to
include only the best hit for each query dna_seg - bidirectional:
Filters the sequence alignment results to include only bidirectional
best hits for each pair of dna_segs - full: The sequence alignment
results are not filtered for best hits
</ul>
<strong><code>--diamond_sensitivity=["SENSITIVITY"]</code></strong>
<ul>
The sensitivity option used when running DIAMOND. This option is ignored
unless DIAMOND was chosen using <code>--comp_format</code>. Must be one
of the following arguments, in order of least to most sensitive:
<code>fast</code>, <code>default</code>, <code>mid-sensitive</code>,
<code>sensitive</code>, <code>more-sensitive</code>,
<code>very-sensitive</code>, <code>ultra-sensitive</code>. (default:
<code>default</code>)
</ul>
<strong><code>--update_positions=[AUTO|TRUE|FALSE]</code></strong>
<ul>
Decides whether or not positions of comparison objects should be updated
to match the positions in the dna_segs. Must be one of:
<code>auto</code>, <code>true</code>, <code>false</code>. If true,
always update, if false, never update. The default is auto, which
updates the positions when <code>--comp_format</code> is DIAMOND or any
of the blastp algorithms.
</ul>
<strong><code>--update_region_plot=[AUTO|TRUE|FALSE]</code></strong>
<ul>
Decides whether or not the region_plot column of dna_segs should be
updated. If <code>true</code>, the region_plot attribute will be given
to each feature that can be linked to a feature that already has the
region_plot attribute. Must be one of: <code>auto</code>,
<code>true</code>, <code>false</code>. If true, always update, if false,
never update. The default is auto, which updates region_plot when
<code>--comp_format</code> is DIAMOND or any of the blastp algorithms.
</ul>
<strong><code>-o [FILE PATH], --out=[FILE PATH]</code></strong>
<ul>
Name of output PDF file. (default: out.pdf)
</ul>
<strong><code>--height=[NUM]</code></strong>
<ul>
PDF height, in inches. (default: auto)
</ul>
<strong><code>--width=[NUM]</code></strong>
<ul>
PDF width, in inches. (default: auto)
</ul>
<strong><code>-t [FILE PATH], --tree=[FILE PATH]</code></strong>
<ul>
Path to newick format tree file with the dna_seg labels (file names) as
leaves.
</ul>
<strong><code>--annotations</code></strong>
<ul>
Gene annotations from the <code>gene</code> column will be plotted.
</ul>
<strong><code>-l [COLUMN], --legend_column=[COLUMN]</code></strong>
<ul>
A dna_seg column to use as the basis for plotting a legend. Must be the
name of a column that appears in all dna_segs or <code>auto</code>. When
<code>auto</code>, it will use the <code>gene</code> column if it is
present, and the <code>name</code> column if it is not, unless
<code>--global_color_scheme</code> is set to <code>uniform</code>, in
which case it will automatically use the same column as the one used to
generate the color scheme. A legend will then be plotted based on the
colors of the dna_seg features, using all of the unique values found in
the chosen column.
</ul>
<strong><code>-g [COLOR SCHEME], --global_color_scheme=[COLOR SCHEME]</code></strong>
<ul>
A color scheme to apply to the dna_segs and/or comparisons. Must be one
of: <code>gradient</code>, <code>uniform</code>,
<code>sequential</code>. - gradient: Calculates color gradients based on
numerical columns. It uses red hues for direct comparisons, and blue
hues for reverse comparisons. It is applied only to comparisons, and
bases the coloring on comparison metrics, namely <code>per_id</code>,
<code>bit_score</code>, and <code>e_value</code> columns, in that order.
- uniform: Generates a color palette based on the possible values a
column can have for all the features that are being displayed. Looks for
a column related to homology, followed by <code>gene</code> and
<code>gene_type</code> if it cannot find any. Applies itself to both
dna_segs and comparisons. - sequential: Takes alterations made on the
colors of features in dna_segs, and transfers them over to the other
dna_segs and comparisons. Only works if changes have been made to at
least 1 dna_seg already. This can be done by using <code>--ids</code>,
for instance.
</ul>
<strong><code>-G [DATASET], --color_scheme_dataset=[DATASET]</code></strong>
<ul>
Which dataset to apply the color scheme to. Only applicable when
<code>--global_color_scheme</code> is set to uniform or sequential. Must
be one of: <code>auto</code>, <code>dna_segs</code>,
<code>comparisons</code>. The default is auto, which will attempt to
automatically determine how best to apply the color scheme based on the
available data.
</ul>
<strong><code>-a [NUM], --alpha_dna_segs=[NUM]</code></strong>
<ul>
A number between 0 (transparent) and 1 (opaque) for the
alpha-transparency level that will be applied to dna_segs.
</ul>
<strong><code>-A [NUM], --alpha_comparisons=[NUM]</code></strong>
<ul>
A number between 0 (transparent) and 1 (opaque) for the
alpha-transparency level that will be applied to comparisons.
</ul>
<strong><code>--dna_seg_scale</code></strong>
<ul>
A scale will be plotted for each dna_seg.
</ul>
<strong><code>-r [NUM], --region_size=[NUM]</code></strong>
<ul>
Neighbourhood size to plot, unless a sequence boundary is found first,
in bp. Setting this argument to 0 will plot the entirety of the
dna_segs. This argument is ignored when xlims are provided via
<code>--xlims_in</code> or <code>--xlims_from_file</code> (default:
10000)
</ul>
<strong><code>-x ["XLIMS_SEG1,XLIMS_SEG2"], --xlims_in=["XLIMS_SEG1,XLIMS_SEG2"]</code></strong>
<ul>
<p>Predefined positions to plot for each dna_seg (and subsegment). If
<code>--xlims_from_file</code> is given, this option is ignored.
dna_segs must be separated by commas, in plotting order, with
subsegments separated by spaces (same format as
<code>--print_xlims</code>). e.g.:</p>
<code>"1 10000 155000 154000,500 10500"</code>
</ul>
<strong><code>-X [FILE PATH], --xlims_from_file=[FILE PATH]</code></strong>
<ul>
<p>Path to a tab-separated file with predefined positions to plot for
each dna_seg (and subsegment) in the same format as that of
<code>--xlims_out</code>. One line per subsegment, in plotting order,
with both positions and their dna_seg label. e.g.:</p>
<table class="table">
<thead><tr class="header">
<th>x0</th>
<th>x1</th>
<th>seg_label</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>13000</td>
<td>dna_seg1</td>
</tr>
<tr class="even">
<td>25000</td>
<td>35000</td>
<td>dna_seg1</td>
</tr>
<tr class="odd">
<td>500</td>
<td>10500</td>
<td>dna_seg2</td>
</tr>
<tr class="even">
<td>25901</td>
<td>35901</td>
<td>dna_seg2</td>
</tr>
</tbody>
</table>
Note: If provided, only dna_segs present in this file will be plotted.
</ul>
<strong><code>-p, --print_xlims</code></strong>
<ul>
Prints out neighbourhood limit positions to the terminal.
</ul>
<strong><code>--xlims_out=[FILE PATH]</code></strong>
<ul>
Name of output xlim file, which will contain the neighbourhood limits
that were used to plot with.
</ul>
<strong><code>-f ["OFFSETS1,OFFSETS2"], --offsets_in=["OFFSETS1,OFFSETS2"]</code></strong>
<ul>
<p>A list of predefined offset values for each dna_seg and subsegment.
If <code>--offsets_from_file</code> is given, this option is ignored.
For each dna_seg 1 value must be provided, or as many values as there
are subsegments for that dna_seg. dna_segs must be separated by commas,
in plotting order, with subsegments separated by spaces. e.g.:</p>
<code>"150 1300,300"</code>
</ul>
<strong><code>-F [FILE PATH], --offsets_from_file=[FILE PATH]</code></strong>
<ul>
<p>Path to a file with predefined offset values for each dna_seg and
subsegment. One line per dna_seg, in plotting order, with subsegments
separated by spaces, e.g.:</p>
<code>150 1300</code> <code>300</code>
</ul>
<strong><code>--print_offsets</code></strong>
<ul>
Prints out offset positions to the terminal.
</ul>
<strong><code>-S [FILE PATH], --dna_segs_out=[FILE PATH]</code></strong>
<ul>
Name of output dna_segs file. If provided, the dna_segs will be saved as
an R object just after reading them in, before any alterations are made
to it. This way, this script can be called again using the same
arguments, except for providing this file to <code>--seg_files</code>,
which should in most cases produce the same result, but without having
to wait for the dna_seg files to be parsed.
</ul>
<strong><code>--workspace_out=[FILE PATH]</code></strong>
<ul>
Name of output file. If provided, all of the arguments that are passed
to main plot function <code><a href="../reference/plot_gene_map.html">plot_gene_map()</a></code> are saved to this
file as well, as an R object that can be loaded in R using the
<code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS()</a></code> function.
</ul>
<strong><code>-n [NUM], --threads=[NUM]</code></strong>
<ul>
Number of threads for parallel processing, used by BLAST and DIAMOND.
(default: 1)
</ul>
<strong><code>-v, --verbose</code></strong>
<ul>
Print out extra information alongside timings, mostly on writing and
reading in files
</ul>
<strong><code>-h, --help</code></strong>
<ul>
Show this help message and exit.
</ul>
</div>
</div>
<div class="section level3">
<h3 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h3>
<div class="section level4">
<h4 id="using-blast-to-create-comparisons-on-the-fly">Using BLAST to create comparisons on the fly<a class="anchor" aria-label="anchor" href="#using-blast-to-create-comparisons-on-the-fly"></a>
</h4>
<p>For this example, we downloaded the GenBank (full) files of the most
recent versions of the four genomes that were used in the dataset found
in <code>data("barto")</code> from NCBI:</p>
<ul>
<li>Bartonella bacilliformis KC583 (NC_008783.1)</li>
<li>Bartonella grahamii as4aup (NC_012846.1)</li>
<li>Bartonella henselae str. Houston-1 (NC_005956.1)</li>
<li>Bartonella quintana str. Toulouse (NC_005955.1)</li>
</ul>
<p>Let’s say we are interested in the <code>addB</code> and
<code>argB</code> genes. <code>genoPlotR</code> will use gene names in
the <code>name</code> column when those are present in the GenBank file,
so we can use these gene names as IDs for this
<code>barto_ids.tab</code> <strong>IDs file</strong>:</p>
<table class="table">
<thead><tr class="header">
<th>id</th>
<th>fill</th>
<th>region_plot</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>addB</td>
<td>red</td>
<td>true</td>
</tr>
<tr class="even">
<td>argB</td>
<td>blue</td>
<td>true</td>
</tr>
</tbody>
</table>
<p>Using this file as input, we can create a basic gene map like
this:</p>
<pre><code>Rscript run_genoPlotR.R \
--seg_files "vignette_barto/GenBank/*" \
--comp_path "vignette_barto/comps" \
--comp_format blastp \
--ids "vignette_barto/barto_ids.tab" \
--out "vignette_barto/ex1_1.pdf"
--width 9</code></pre>
<p>This will use <code>blastp</code> to create comparisons and store the
resulting files in <code>vignette_barto/comps</code>. If we were to make
another gene map with the same input genomes, we would not need to run
BLAST a second time, even if we were to look at a different region of
these genomes.</p>
<p><img src="ex1_1.png" width="95%"></p>
</div>
<div class="section level4">
<h4 id="customization">Customization<a class="anchor" aria-label="anchor" href="#customization"></a>
</h4>
<p>The comparisons are a bit difficult to track and the figure is fairly
bare-bones in general. So, let’s run it again, with a few visual
customization options added in:</p>
<pre><code>Rscript run_genoPlotR.R \
--seg_files "vignette_barto/GenBank/*" \
--comp_path "vignette_barto/comps" \
--comp_format blastp \
--ids "vignette_barto/barto_ids.tab" \
--out "vignette_barto/ex1_2.pdf" \
--width 9 \
--annotations \
--alpha_comparisons 0.5 \
--global_color_scheme sequential \
--print_xlims</code></pre>
<p>Instead of showing all BLAST results, we can filter them using
<code>--comp_mode</code>. This option can filter the BLAST results by
taking only the best hit per query (<code>besthit</code>), or only
bidirectional best hits (<code>bidirectional</code>). Picking
<code>bidirectional</code> does double the amount of BLAST results
generated, as the best hits for the BLAST results in 1 direction now
have to match those of the other direction. Adding
<code>--annotations</code> will automatically add annotations,
displaying gene names, or locus tags when these are not available. The
<code>--alpha_comparisons 0.5</code> option adds transparency to the
comparisons. easier to parse visually. The
<code>--global_color_scheme sequential</code> option transfers over the
colors we added earlier to the comparisons they are connected to. This
results in the following figure:</p>
<p><img src="ex1_2.png" width="95%"></p>
<p>Note that we also added <code>--print_xlims</code>. This printed out
the <code>xlims</code>, the start and end coordinates for each
subsegment and DNA segment.</p>
<pre><code>1368895 1389810 1412813 1435966,16890 40019 79755 100670,21410 44542 84165 105080,19042 42171 74542 95457</code></pre>
<p>In the examples above, each DNA segment has 2 subsegments, so 4
coordinates per DNA segment, with DNA segments separated by commas.
Because we chose to print these <code>xlims</code>, we can now
manipulate them. For example, we can enter the coordinates in reverse
order for <em>B_bacciliformis</em>, inverting that entire DNA segment.
To do so, we can use the <code>--xlims_in</code> option, which allows us
to manually enter in coordinates in the same format as the output of
<code>--print_xlims</code>:</p>
<pre><code>Rscript run_genoPlotR.R \
--seg_files "vignette_barto/GenBank/*" \
--comp_path "vignette_barto/comps" \
--comp_format blastp \
--ids "vignette_barto/barto_ids.tab" \
--out "vignette_barto/ex1_3.pdf" \
--width 9 \
--annotations \
--alpha_comparisons 0.5 \
--global_color_scheme sequential \
--xlims_in "1435966 1412813 1389810 1368895,16890 40019 79755 100670,21410 44542 84165 105080,19042 42171 74542 95457"</code></pre>
<p><img src="ex1_3.png" width="95%"></p>
<p>Selecting <code>uniform</code> for the
<code>--global_color_scheme</code> option instead of
<code>sequential</code> automatically assigns colors based on the data
is available. In this example, it uses the gene names. Hypothetical
proteins are given a single, separate color instead. This results in the
following plot:</p>
<p><img src="ex1_4.png" width="95%"></p>
<p>The <code>--legend_column</code> option will generate a legend based
on the values found in a <code>dna_seg</code> column of choice.
Alternatively, you can set it to <code>auto</code>. If
<code>--global_color_scheme uniform</code> was used, the legend will
automatically be based on the same column that was used to generate the
color scheme. In all other cases, picking <code>auto</code> defaults to
the <code>gene</code> column, or the <code>name</code> column if the
<code>gene</code> column is not present. So, by adding
<code>--legend_column auto</code> to the previous plot we get this:</p>
<p><img src="ex1_5.png" width="95%"></p>
</div>
<div class="section level4">
<h4 id="visualing-cluster-results">Visualing cluster results<a class="anchor" aria-label="anchor" href="#visualing-cluster-results"></a>
</h4>
<p>In addition to sequence alignments from BLAST and DIAMOND,
<code>run_genoPlotR.R</code> also supports the use of sequence
clustering tools to create the comparisons between DNA segments. Support
is included for the output formats used by OrthoMCL, Orthofinder, and
mmseqs2. This method requires a bit more preparation, as the results
must be created beforehand. However, it does offer certain advantages as
well.</p>
<p>For example, we can use DIAMOND to cluster the sequences. This
requires FASTA files for each of the input DNA segments. In the previous
examples, BLAST was used to create the comparisons, which also generates
the necessary FASTA files. The <code>genoPlotR</code> includes
<code><a href="../reference/genbank_to_fasta.html">genbank_to_fasta()</a></code> and <code><a href="../reference/dna_seg_to_fasta.html">dna_seg_to_fasta()</a></code>
functions to create FASTA files directly from GenBank files or
<code>dna_seg</code> objects, respectively. Assuming the FASTA files
have been made, you could use a command like this to concatenate them
all:</p>
<pre><code>cat vignette_barto/fasta/*.fa &gt; vignette_barto/concat.fa</code></pre>
<p>The resulting file can then be used for DIAMOND:</p>
<pre><code>diamond cluster -d "vignette_barto/concat.fa" -o "vignette_barto/clusters.tab"</code></pre>
<p>We can now use this as input for <code>run_genoPlotR.R</code> by
selecting this file with <code>--comp_path</code> and setting
<code>--comp_format</code> to <code>mmseqs2</code>.
<strong>IMPORTANT:</strong> Entering <code>--comp_format diamond</code>
would result in using DIAMOND for pairwise sequence alignments, not
clustering. The clustering results from DIAMOND are in the same format
as those produced by mmseqs2, which is why we select that option
instead. One of the advantages of using clustering is that the order of
the DNA segments can be freely changed, as the comparisons are no longer
created from pairwise sequence alignments. Another advantage is that all
<code>dna_seg</code> features that are in the same cluster are
automatically given the same colors when the uniform color scheme is
chosen. In this example, the comparisons made are the same as those from
bidirectional best hits from BLAST, but clustering results are often
more reliable.</p>
<pre><code>Rscript run_genoPlotR.R \
--seg_files "vignette_barto/GenBank/*" \
--comp_path "vignette_barto/clusters.tab" \
--comp_format mmseqs2 \
--ids "vignette_barto/barto_ids.tab" \
--out "vignette_barto/ex1_5.pdf" \
--width 9 \
--annotations \
--alpha_comparisons 0.5 \
--global_color_scheme uniform \
--xlims_in "1435966 1412813 1389810 1368895,16890 40019 79755 100670,21410 44542 84165 105080,19042 42171 74542 95457"</code></pre>
<p><img src="ex1_6.png" width="95%"></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mike Puijk.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
